## Install Bokeh visualization package
We pip install this first because it restarts the Python interpreter. If this is done later in the notebook, all the variables prior to running the pip install will be lost.

In [0]:
%pip install bokeh

Python interpreter will be restarted.
Collecting bokeh
 Downloading bokeh-2.2.3.tar.gz (8.8 MB)
Collecting PyYAML>=3.10
 Downloading PyYAML-5.3.1.tar.gz (269 kB)
Requirement already satisfied: python-dateutil>=2.1 in /databricks/python3/lib/python3.7/site-packages (from bokeh) (2.8.1)
Collecting Jinja2>=2.7
 Downloading Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
Requirement already satisfied: numpy>=1.11.3 in /databricks/python3/lib/python3.7/site-packages (from bokeh) (1.18.1)
Collecting pillow>=7.1.0
 Downloading Pillow-8.0.1-cp37-cp37m-manylinux1_x86_64.whl (2.2 MB)
Collecting packaging>=16.8
 Downloading packaging-20.4-py2.py3-none-any.whl (37 kB)
Requirement already satisfied: tornado>=5.1 in /databricks/python3/lib/python3.7/site-packages (from bokeh) (6.0.3)
Collecting typing_extensions>=3.7.4
 Downloading typing_extensions-3.7.4.3-py3-none-any.whl (22 kB)
Requirement already satisfied: six>=1.5 in /databricks/python3/lib/python3.7/site-packages (from python-dateutil>=2.1->bokeh) (1.14.0)
Collecting MarkupSafe>=0.23
 Downloading MarkupSafe-1.1.1-cp37-cp37m-manylinux1_x86_64.whl (27 kB)
Requirement already satisfied: pyparsing>=2.0.2 in /databricks/python3/lib/python3.7/site-packages (from packaging>=16.8->bokeh) (2.4.6)
Building wheels for collected packages: bokeh, PyYAML
 Building wheel for bokeh (setup.py): started
 Building wheel for bokeh (setup.py): finished with status 'done'
 Created wheel for bokeh: filename=bokeh-2.2.3-py3-none-any.whl size=9296309 sha256=279b94fe5a2ffbb23053540e5384a37e708df03f641afc98ca15a8931f3baf8f
 Stored in directory: /root/.cache/pip/wheels/0c/6a/ac/5de7ec1983d76fa1243106008b70df211dceb5ed643ad15699
 Building wheel for PyYAML (setup.py): started
 Building wheel for PyYAML (setup.py): finished with status 'done'
 Created wheel for PyYAML: filename=PyYAML-5.3.1-cp37-cp37m-linux_x86_64.whl size=44619 sha256=3d982cec826c300912f77ea3883ee64ce344d6adb4ee1f6513c0906c366de787
 Stored in directory: /root/.cache/pip/wheels/5e/03/1e/e1e954795d6f35dfc7b637fe2277bff021303bd9570ecea653
Successfully built bokeh PyYAML
Installing collected packages: PyYAML, MarkupSafe, Jinja2, pillow, packaging, typing-extensions, bokeh
Successfully installed Jinja2-2.11.2 MarkupSafe-1.1.1 PyYAML-5.3.1 bokeh-2.2.3 packaging-20.4 pillow-8.0.1 typing-extensions-3.7.4.3
Python interpreter will be restarted.

# Begin Spark Session

In [0]:
from pyspark.sql import SparkSession
from pyspark import SparkContext


spark = SparkSession.builder.appName('connecticut').getOrCreate()

from pyspark.sql import functions as F

### Read Data into Dataframe

In [0]:
rdd = sc.textFile("dbfs:/FileStore/shared_uploads/kariboalap@uni.coventry.ac.uk/Real_Estate_Sales_2001_2018__1_.csv")
rdd.map(lambda row:row.split(',')).take(2)

Out[2]: [['SerialNumber',
 'ListYear',
 'DateRecorded',
 'Town',
 'Address',
 'AssessedValue',
 'SaleAmount',
 'SalesRatio',
 'PropertyType',
 'ResidentialType',
 'NonUseCode',
 'Remarks',
 'NumberOfBuildings'],
 ['17001',
 '2017',
 '10/02/2017',
 'Andover',
 '27 HENDEE RD',
 '58400',
 '82000',
 '0.712',
 '',
 'Single Family',
 '',
 '',
 '']]

In [0]:
# Read the data using Dataframes method
conn = spark.read.format("csv").load("dbfs:/FileStore/shared_uploads/kariboalap@uni.coventry.ac.uk/Real_Estate_Sales_2001_2018__1_.csv", inferSchema=True, header=True)

## 1. Exploration
#### 1.1. Display top 5 rows

In [0]:
# conn.show(n)     <------- General syntax to show first n number of rows
conn.display()

SerialNumber ListYear DateRecorded Town Address AssessedValue SaleAmount SalesRatio PropertyType ResidentialType NonUseCode Remarks NumberOfBuildings 17001 2017 10/02/2017 Andover 27 HENDEE RD 58400 82000.0 0.712 null Single Family null null null 40043 2004 05/19/2005 Andover BURNAP BRK RD 850 50000.0 1.7 Vacant Land NA 28 NA null 30070 2003 08/02/2004 Andover ROUTE 6 157300 175000.0 89.88 Vacant Land NA 0 NA null 40036 2004 04/29/2005 Andover ROUTE 6 22400 20000.0 112 Vacant Land NA 25 NA null 170003 2017 10/02/2017 Branford 22-24 SILVER ST 205500 357500.0 0.575 null Two Family null null null 50249 2005 07/05/2006 Ansonia 13 MCINTOSH LN 0 197400.0 0 Condo NA 7 NA null 50125 2005 02/16/2006 Ansonia 22 WESTFIELD AVE 5 & 331100 375000.0 88.3 Condo NA 0 NA null 170007 2017 10/02/2017 Groton 174 ANN AVE 158410 192000.0 0.825 null Single Family null null null 50059 2005 05/17/2006 Ashford 52 PERRY HL RD 1196070 3700000.0 32.3 Apartments NA 0 NA null 140042 2014 07/22/2015 Ashford ASHFORD LAKE DR 3900 2500.0 1.56 Vacant Land NA NA NA null 900009 2009 12/09/2009 Ashford HILLSIDE RD 64000 160000.0 0.4 Vacant Land NA NA NA null 170306 2017 10/02/2017 Milford 60 PAMELA DR 197070 288000.0 0.684 null Single Family null null null 20384 2002 06/16/2003 Avon 1 WILLS WALK 334750 577500.0 57.965367965368 Condo NA 0 NA null 120148 2012 03/04/2013 Avon 11 CHESTNUT DRIVE 125870 125000.0 1.00696 Commercial NA NA 0 null 40478 2004 07/29/2005 Avon 11 CLOVER CT 109550 229500.0 47.7342047930283 Condo NA 0 NA null 20287 2002 04/23/2003 Avon 11 WILLOW LN 77000 147500.0 52.2033898305085 Condo NA 0 NA null 20345 2002 05/22/2003 Avon 12 WYNDEMERE 38500 546987.0 7.03855850321854 Condo NA 7 NA null 100064 2010 12/16/2010 Avon 16 GATEWOOD 354310 550000.0 0.6442 Condo NA NA NA null 40056 2004 11/04/2004 Avon 2 DOVE CIR 105350 175000.0 60.2 Condo NA 0 NA null 60014 2006 10/11/2006 Avon 2 SYCAMORE LN 102900 250000.0 0.4116 Condo NA NA addl remarks null 40163 2004 01/18/2005 Avon 2 TEMPLETON CT 307020 430000.0 71.4 Condo NA 0 NA null 20194 2002 02/18/2003 Avon 3 BRIDLE PATH 173880 380000.0 45.7578947368421 Condo NA 0 NA null 17006 2017 10/03/2017 Rocky Hill 85A BROOKWOOD DRIVE 77980 90000.0 0.866 null Condo Family null null null 10408 2001 07/17/2002 Avon 4 SYCAMORE LN 87080 165000.0 52.7757575757576 Condo NA 0 NA null 170005 2017 10/04/2017 Madison 28 IVY ISLAND DR 242400 437750.0 0.554 null Single Family null null null 20207 2002 02/28/2003 Avon 48 BYRON DR 125860 235000.0 53.5574468085106 Condo NA 0 NA null 10340 2001 06/13/2002 Avon 50 SCHOOL ST 42000 10000.0 420 Vacant Land NA 24 NA null 50288 2005 06/01/2006 Avon 6 GREENWICH LN 170790 339500.0 50.3 Condo NA 0 NA null 50257 2005 05/01/2006 Avon 7 ESSEX CT 153240 284000.0 54 Condo NA 0 NA null 20236 2002 03/18/2003 Avon 7 MADISON LN 163910 327500.0 50.0488549618321 Condo NA 0 NA null 40404 2004 06/27/2005 Avon 7 MADISON LN 189700 350000.0 54.2 Condo NA 0 NA null 170002 2017 10/05/2017 Winchester 109 BLUE ST 144830 235000.0 0.616 null Single Family null null null 40384 2004 06/20/2005 Avon 70 FOX HOLW 233180 399000.0 58.4411027568922 Condo NA 0 NA null 20448 2002 07/01/2003 Avon 9 WILLOW LN 87080 189000.0 46.0740740740741 Condo NA 0 NA null 170006 2017 10/06/2017 Ledyard 12 REUVEN DR 213710 290000.0 0.737 null Single Family null null null 170010 2017 10/10/2017 Canton 75 BART DRIVE 517810 630000.0 0.822 null Single Family null null null 30284 2003 06/28/2004 Berlin 92 HAWTHORNE DR 13600 245725.0 5.53 Condo NA 7 NA null 170028 2017 10/10/2017 Trumbull 395 DAYTON RD 224070 374000.0 0.599 null Single Family null null null 70072 2007 05/12/2008 Beacon Falls 6 FIELDSTONE LN 0 508955.0 0 Condo NA 7 NA null 50320 2005 08/01/2006 Berlin 159 SCARBOROUGH FARE 183700 362000.0 50.7 Condo NA 0 NA null 900194 2009 06/17/2010 Berlin 17 IRIS WAY 0 362031.0 0 Condo NA 07 NA null 170013 2017 10/11/2017 Rocky Hill 28A ROBBINS LANE 45150 70000.0 0.645 null Condo Family null null null 100185 2010 06/01/2011 Berlin 501 BERLIN TPKE 

#### 1.2.  View datatypes of all columns

In [0]:
conn.printSchema()

root
-- SerialNumber: integer (nullable = true)
-- ListYear: integer (nullable = true)
-- DateRecorded: string (nullable = true)
-- Town: string (nullable = true)
-- Address: string (nullable = true)
-- AssessedValue: integer (nullable = true)
-- SaleAmount: double (nullable = true)
-- SalesRatio: string (nullable = true)
-- PropertyType: string (nullable = true)
-- ResidentialType: string (nullable = true)
-- NonUseCode: string (nullable = true)
-- Remarks: string (nullable = true)
-- NumberOfBuildings: string (nullable = true)

#### 1.3.   Statistical Summary of the dataset
We will only be doing this for numerical columns as there is no need for statistical description of string columns

In [0]:
conn.describe().display()

summary SerialNumber ListYear DateRecorded Town Address AssessedValue SaleAmount SalesRatio PropertyType ResidentialType NonUseCode Remarks NumberOfBuildings count 912302 912302 912294 912302 912299 911997 881509 912301 815904 912301 847221 837966 50713 mean 159097.28515009285 2008.4892349244 null null 67823.8431372549 268451.45824821794 362152.8324535428 713.9486970041747 null null 4.778324718317427 2969.040750635723 0.77301775147929 stddev 2578058.8060479234 5.56863296484261 null null 105734.86703455442 1644922.29602784 2006128.6417481834 133131.20656566767 null null 7.8861093739395445 30423.43072081334 0.7279760144235242 min 0 2001 01/01/2002 Andover "RT 197 PARCEL ""B"" NORTH" 0 0.0 0 10 Mill Forest Apartments 0 "160-6-11 SELLER IS ""YMCA""" ESTATE SALE" max 1710011174 2018 12/31/2018 Woodstock parking space only 881510000 9.4094E8 NA Vacant Land Vacant Land NA zoning changed from residential to commercial after Oct 1. NO BSMNTM NO BACK YARD

In [0]:
conn.describe(['ListYear','AssessedValue','SaleAmount',]).show()

+-------+----------------+------------------+------------------+
summary| ListYear| AssessedValue| SaleAmount|
+-------+----------------+------------------+------------------+
 count| 912302| 911997| 881509|
 mean| 2008.4892349244|268451.45824821794| 362152.8324535428|
 stddev|5.56863296484261| 1644922.29602784|2006128.6417481834|
 min| 2001| 0| 0.0|
 max| 2018| 881510000| 9.4094E8|
+-------+----------------+------------------+------------------+

#### 1.4  Show Unique Values

* ##### Towns

In [0]:
# Unique towns
conn.select('Town').distinct().show(5)

#Number of unique towns
conn.select(F.countDistinct("Town")).show()

+--------+
 Town|
+--------+
 Ashford|
 Andover|
 Groton|
Branford|
 Milford|
+--------+
only showing top 5 rows

+--------------------+
count(DISTINCT Town)|
+--------------------+
 169|
+--------------------+

In [0]:
# Which towns are the most frequently occuring (Top 10)
conn.groupBy("Town").count().orderBy('count', ascending=False).show(10)

+-------------+-----+
 Town|count|
+-------------+-----+
 Bridgeport|34486|
 Stamford|29462|
 Waterbury|24275|
 Norwalk|21474|
 New Haven|20463|
West Hartford|18322|
 Hartford|17705|
 Danbury|17460|
 Milford|16407|
 Meriden|16273|
+-------------+-----+
only showing top 10 rows

* ##### Property Type

In [0]:
# Unique property types
conn.select('PropertyType').distinct().show()

#Number of unique property types
conn.select(F.countDistinct("PropertyType")).show()

+--------------+
 PropertyType|
+--------------+
 Apartments|
 Vacant Land|
 NA|
 Residential|
 null|
 Industrial|
 Condo|
Public Utility|
10 Mill Forest|
 Commercial|
+--------------+

+----------------------------+
count(DISTINCT PropertyType)|
+----------------------------+
 9|
+----------------------------+

In [0]:
# Which Property types are the most frequently occuring (Top 10)

conn.groupBy("PropertyType").count().orderBy('count', ascending=False).show(10)

+--------------+------+
 PropertyType| count|
+--------------+------+
 Residential|602891|
 Condo|120527|
 null| 96398|
 Vacant Land| 46356|
 Commercial| 32262|
 Apartments| 5949|
 Industrial| 4181|
 NA| 3586|
Public Utility| 141|
10 Mill Forest| 11|
+--------------+------+

* ##### Residential Type

In [0]:
# Unique residential types
conn.select('ResidentialType').distinct().show()

#Number of unique residential types
conn.select(F.countDistinct("ResidentialType")).show()

+---------------+
ResidentialType|
+---------------+
 Four Family|
 Apartments|
 Vacant Land|
 NA|
 Residential|
 Industrial|
 Two Family|
 Single Family|
 Multi Family|
 Public Utility|
 Commercial|
 Three Family|
 Condo Family|
 null|
+---------------+

+-------------------------------+
count(DISTINCT ResidentialType)|
+-------------------------------+
 13|
+-------------------------------+

In [0]:
# Which Residential types are the most frequently occuring (Top 10)
conn.groupBy("ResidentialType").count().orderBy('count', ascending=False).show(10)

+---------------+------+
ResidentialType| count|
+---------------+------+
 Single Family|614927|
 NA|213196|
 Two Family| 39037|
 Three Family| 19453|
 Condo Family| 16180|
 Vacant Land| 3555|
 Commercial| 3297|
 Four Family| 1332|
 Apartments| 543|
 Industrial| 404|
+---------------+------+
only showing top 10 rows

#### 1.5  Total listings for each year

Here we can quickly see what year had the most listings. Gives us an idea of the size of the real estate market by year

In [0]:
# Total listings for each year
conn.groupBy("ListYear").count().orderBy('ListYear', ascending=True).show()

+--------+------+
ListYear| count|
+--------+------+
 2001| 59584|
 2002|106068|
 2003| 64239|
 2004| 84056|
 2005| 61602|
 2006| 48775|
 2007| 35616|
 2008| 32734|
 2009| 42508|
 2010| 27755|
 2011| 31065|
 2012| 35973|
 2013| 39943|
 2014| 49563|
 2015| 46651|
 2016| 49773|
 2017| 45692|
 2018| 50705|
+--------+------+

While these tables do a good job of quickly describing the dataset, the information is better communicated using visualizations. <br>
Before the dataset can be visualized, it will need to undergo wrangling to ensure proper formatting and also enrichment.

## 2. Wrangling

Let us look at the dataframe again for reference

In [0]:
conn.display()

SerialNumber ListYear DateRecorded Town Address AssessedValue SaleAmount SalesRatio PropertyType ResidentialType NonUseCode Remarks NumberOfBuildings 17001 2017 10/02/2017 Andover 27 HENDEE RD 58400 82000.0 0.712 null Single Family null null null 40043 2004 05/19/2005 Andover BURNAP BRK RD 850 50000.0 1.7 Vacant Land NA 28 NA null 30070 2003 08/02/2004 Andover ROUTE 6 157300 175000.0 89.88 Vacant Land NA 0 NA null 40036 2004 04/29/2005 Andover ROUTE 6 22400 20000.0 112 Vacant Land NA 25 NA null 170003 2017 10/02/2017 Branford 22-24 SILVER ST 205500 357500.0 0.575 null Two Family null null null 50249 2005 07/05/2006 Ansonia 13 MCINTOSH LN 0 197400.0 0 Condo NA 7 NA null 50125 2005 02/16/2006 Ansonia 22 WESTFIELD AVE 5 & 331100 375000.0 88.3 Condo NA 0 NA null 170007 2017 10/02/2017 Groton 174 ANN AVE 158410 192000.0 0.825 null Single Family null null null 50059 2005 05/17/2006 Ashford 52 PERRY HL RD 1196070 3700000.0 32.3 Apartments NA 0 NA null 140042 2014 07/22/2015 Ashford ASHFORD LAKE DR 3900 2500.0 1.56 Vacant Land NA NA NA null 900009 2009 12/09/2009 Ashford HILLSIDE RD 64000 160000.0 0.4 Vacant Land NA NA NA null 170306 2017 10/02/2017 Milford 60 PAMELA DR 197070 288000.0 0.684 null Single Family null null null 20384 2002 06/16/2003 Avon 1 WILLS WALK 334750 577500.0 57.965367965368 Condo NA 0 NA null 120148 2012 03/04/2013 Avon 11 CHESTNUT DRIVE 125870 125000.0 1.00696 Commercial NA NA 0 null 40478 2004 07/29/2005 Avon 11 CLOVER CT 109550 229500.0 47.7342047930283 Condo NA 0 NA null 20287 2002 04/23/2003 Avon 11 WILLOW LN 77000 147500.0 52.2033898305085 Condo NA 0 NA null 20345 2002 05/22/2003 Avon 12 WYNDEMERE 38500 546987.0 7.03855850321854 Condo NA 7 NA null 100064 2010 12/16/2010 Avon 16 GATEWOOD 354310 550000.0 0.6442 Condo NA NA NA null 40056 2004 11/04/2004 Avon 2 DOVE CIR 105350 175000.0 60.2 Condo NA 0 NA null 60014 2006 10/11/2006 Avon 2 SYCAMORE LN 102900 250000.0 0.4116 Condo NA NA addl remarks null 40163 2004 01/18/2005 Avon 2 TEMPLETON CT 307020 430000.0 71.4 Condo NA 0 NA null 20194 2002 02/18/2003 Avon 3 BRIDLE PATH 173880 380000.0 45.7578947368421 Condo NA 0 NA null 17006 2017 10/03/2017 Rocky Hill 85A BROOKWOOD DRIVE 77980 90000.0 0.866 null Condo Family null null null 10408 2001 07/17/2002 Avon 4 SYCAMORE LN 87080 165000.0 52.7757575757576 Condo NA 0 NA null 170005 2017 10/04/2017 Madison 28 IVY ISLAND DR 242400 437750.0 0.554 null Single Family null null null 20207 2002 02/28/2003 Avon 48 BYRON DR 125860 235000.0 53.5574468085106 Condo NA 0 NA null 10340 2001 06/13/2002 Avon 50 SCHOOL ST 42000 10000.0 420 Vacant Land NA 24 NA null 50288 2005 06/01/2006 Avon 6 GREENWICH LN 170790 339500.0 50.3 Condo NA 0 NA null 50257 2005 05/01/2006 Avon 7 ESSEX CT 153240 284000.0 54 Condo NA 0 NA null 20236 2002 03/18/2003 Avon 7 MADISON LN 163910 327500.0 50.0488549618321 Condo NA 0 NA null 40404 2004 06/27/2005 Avon 7 MADISON LN 189700 350000.0 54.2 Condo NA 0 NA null 170002 2017 10/05/2017 Winchester 109 BLUE ST 144830 235000.0 0.616 null Single Family null null null 40384 2004 06/20/2005 Avon 70 FOX HOLW 233180 399000.0 58.4411027568922 Condo NA 0 NA null 20448 2002 07/01/2003 Avon 9 WILLOW LN 87080 189000.0 46.0740740740741 Condo NA 0 NA null 170006 2017 10/06/2017 Ledyard 12 REUVEN DR 213710 290000.0 0.737 null Single Family null null null 170010 2017 10/10/2017 Canton 75 BART DRIVE 517810 630000.0 0.822 null Single Family null null null 30284 2003 06/28/2004 Berlin 92 HAWTHORNE DR 13600 245725.0 5.53 Condo NA 7 NA null 170028 2017 10/10/2017 Trumbull 395 DAYTON RD 224070 374000.0 0.599 null Single Family null null null 70072 2007 05/12/2008 Beacon Falls 6 FIELDSTONE LN 0 508955.0 0 Condo NA 7 NA null 50320 2005 08/01/2006 Berlin 159 SCARBOROUGH FARE 183700 362000.0 50.7 Condo NA 0 NA null 900194 2009 06/17/2010 Berlin 17 IRIS WAY 0 362031.0 0 Condo NA 07 NA null 170013 2017 10/11/2017 Rocky Hill 28A ROBBINS LANE 45150 70000.0 0.645 null Condo Family null null null 100185 2010 06/01/2011 Berlin 501 BERLIN TPKE 

Each column is examined with the purpose of converting them to useful formats and datatypes.

#### 2.1. DateRecorded

This column shows when the property sale was recorded. **I will be assuming** this to be the **sale date** for each entry.

It is given in the format: **MM/DD/YYYY**, but it will be more useful to split this into **sale month** and **sale year**.

In [0]:
from pyspark.sql.functions import col, unix_timestamp, to_date

# First convert the DateRecorded column datatype from string to date 
conn1 = conn.withColumn('SaleDate', to_date(unix_timestamp(col('DateRecorded'), 'MM/dd/yyyy').cast("timestamp")))
conn1.select('DateRecorded','SaleDate').show(5)
conn1.select('DateRecorded','SaleDate').printSchema()                  

+------------+----------+
DateRecorded| SaleDate|
+------------+----------+
 10/02/2017|2017-10-02|
 05/19/2005|2005-05-19|
 08/02/2004|2004-08-02|
 04/29/2005|2005-04-29|
 10/02/2017|2017-10-02|
+------------+----------+
only showing top 5 rows

root
-- DateRecorded: string (nullable = true)
-- SaleDate: date (nullable = true)

In [0]:
# Now extract Year from SaleDate

from pyspark.sql.functions import year
from pyspark.sql.functions import to_date

conn1 = conn1.withColumn('SaleYear',year(conn1.SaleDate))
conn1.select('SaleDate','SaleYear').show(5)


+----------+--------+
 SaleDate|SaleYear|
+----------+--------+
2017-10-02| 2017|
2005-05-19| 2005|
2004-08-02| 2004|
2005-04-29| 2005|
2017-10-02| 2017|
+----------+--------+
only showing top 5 rows

In [0]:
# Extract Sale Month
from pyspark.sql.functions import month
conn1 = conn1.withColumn('SaleMonth',month(conn1.SaleDate))
conn1.select('SaleDate','SaleMonth').show(5)

+----------+---------+
 SaleDate|SaleMonth|
+----------+---------+
2017-10-02| 10|
2005-05-19| 5|
2004-08-02| 8|
2005-04-29| 4|
2017-10-02| 10|
+----------+---------+
only showing top 5 rows

In [0]:
conn1.printSchema()

root
-- SerialNumber: integer (nullable = true)
-- ListYear: integer (nullable = true)
-- DateRecorded: string (nullable = true)
-- Town: string (nullable = true)
-- Address: string (nullable = true)
-- AssessedValue: integer (nullable = true)
-- SaleAmount: double (nullable = true)
-- SalesRatio: string (nullable = true)
-- PropertyType: string (nullable = true)
-- ResidentialType: string (nullable = true)
-- NonUseCode: string (nullable = true)
-- Remarks: string (nullable = true)
-- NumberOfBuildings: string (nullable = true)
-- SaleDate: date (nullable = true)
-- SaleYear: integer (nullable = true)
-- SaleMonth: integer (nullable = true)

#### 2.2. AssessedValue and SaleAmount

There two columns represent the **amount the property was valued at**, perhaps at the time of listing, and the **amount it was eventually sold for**.

The difference between these values can be taken as an additional variable, as it contains information on the accuracy of property valuations. 

A new column `PriceDifference` is hence created.

In [0]:
conn2 = conn1.withColumn('PriceDifference', conn1.AssessedValue - conn1.SaleAmount)
conn2.select('AssessedValue','SaleAmount','PriceDifference').show(5)


+-------------+----------+---------------+
AssessedValue|SaleAmount|PriceDifference|
+-------------+----------+---------------+
 58400| 82000.0| -23600.0|
 850| 50000.0| -49150.0|
 157300| 175000.0| -17700.0|
 22400| 20000.0| 2400.0|
 205500| 357500.0| -152000.0|
+-------------+----------+---------------+
only showing top 5 rows

#### 2.3. Town and Address

In [0]:
conn2.select('Town', 'Address').show(5)

+--------+---------------+
 Town| Address|
+--------+---------------+
 Andover| 27 HENDEE RD|
 Andover| BURNAP BRK RD|
 Andover| ROUTE 6|
 Andover| ROUTE 6|
Branford|22-24 SILVER ST|
+--------+---------------+
only showing top 5 rows

We will attempt to geocode some of the addresses to coordinates in order to view price trends on a map.  This will come much later during visualizations.

## 3  Cleaning
#### 3.1. Drop Columns

Now, columns not holding any useful information will be dropped. In the same command, the columns will also be reordered

In [0]:
# Drop unneccesary columns. Here we simply select the columns, but leaving out the 'Remarks','NoneUseCode' and 'No of Buildings' columns 
conn3 = conn2.select('SerialNumber','ListYear','SaleYear','SaleMonth',
                     'AssessedValue','SaleAmount','PriceDifference',
                     'PropertyType','ResidentialType','Address','Town')
conn3.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 17001 2017 2017 10 58400 82000.0 -23600.0 null Single Family 27 HENDEE RD Andover 40043 2004 2005 5 850 50000.0 -49150.0 Vacant Land NA BURNAP BRK RD Andover 30070 2003 2004 8 157300 175000.0 -17700.0 Vacant Land NA ROUTE 6 Andover 40036 2004 2005 4 22400 20000.0 2400.0 Vacant Land NA ROUTE 6 Andover 170003 2017 2017 10 205500 357500.0 -152000.0 null Two Family 22-24 SILVER ST Branford 50249 2005 2006 7 0 197400.0 -197400.0 Condo NA 13 MCINTOSH LN Ansonia 50125 2005 2006 2 331100 375000.0 -43900.0 Condo NA 22 WESTFIELD AVE 5 & Ansonia 170007 2017 2017 10 158410 192000.0 -33590.0 null Single Family 174 ANN AVE Groton 50059 2005 2006 5 1196070 3700000.0 -2503930.0 Apartments NA 52 PERRY HL RD Ashford 140042 2014 2015 7 3900 2500.0 1400.0 Vacant Land NA ASHFORD LAKE DR Ashford 900009 2009 2009 12 64000 160000.0 -96000.0 Vacant Land NA HILLSIDE RD Ashford 170306 2017 2017 10 197070 288000.0 -90930.0 null Single Family 60 PAMELA DR Milford 20384 2002 2003 6 334750 577500.0 -242750.0 Condo NA 1 WILLS WALK Avon 120148 2012 2013 3 125870 125000.0 870.0 Commercial NA 11 CHESTNUT DRIVE Avon 40478 2004 2005 7 109550 229500.0 -119950.0 Condo NA 11 CLOVER CT Avon 20287 2002 2003 4 77000 147500.0 -70500.0 Condo NA 11 WILLOW LN Avon 20345 2002 2003 5 38500 546987.0 -508487.0 Condo NA 12 WYNDEMERE Avon 100064 2010 2010 12 354310 550000.0 -195690.0 Condo NA 16 GATEWOOD Avon 40056 2004 2004 11 105350 175000.0 -69650.0 Condo NA 2 DOVE CIR Avon 60014 2006 2006 10 102900 250000.0 -147100.0 Condo NA 2 SYCAMORE LN Avon 40163 2004 2005 1 307020 430000.0 -122980.0 Condo NA 2 TEMPLETON CT Avon 20194 2002 2003 2 173880 380000.0 -206120.0 Condo NA 3 BRIDLE PATH Avon 17006 2017 2017 10 77980 90000.0 -12020.0 null Condo Family 85A BROOKWOOD DRIVE Rocky Hill 10408 2001 2002 7 87080 165000.0 -77920.0 Condo NA 4 SYCAMORE LN Avon 170005 2017 2017 10 242400 437750.0 -195350.0 null Single Family 28 IVY ISLAND DR Madison 20207 2002 2003 2 125860 235000.0 -109140.0 Condo NA 48 BYRON DR Avon 10340 2001 2002 6 42000 10000.0 32000.0 Vacant Land NA 50 SCHOOL ST Avon 50288 2005 2006 6 170790 339500.0 -168710.0 Condo NA 6 GREENWICH LN Avon 50257 2005 2006 5 153240 284000.0 -130760.0 Condo NA 7 ESSEX CT Avon 20236 2002 2003 3 163910 327500.0 -163590.0 Condo NA 7 MADISON LN Avon 40404 2004 2005 6 189700 350000.0 -160300.0 Condo NA 7 MADISON LN Avon 170002 2017 2017 10 144830 235000.0 -90170.0 null Single Family 109 BLUE ST Winchester 40384 2004 2005 6 233180 399000.0 -165820.0 Condo NA 70 FOX HOLW Avon 20448 2002 2003 7 87080 189000.0 -101920.0 Condo NA 9 WILLOW LN Avon 170006 2017 2017 10 213710 290000.0 -76290.0 null Single Family 12 REUVEN DR Ledyard 170010 2017 2017 10 517810 630000.0 -112190.0 null Single Family 75 BART DRIVE Canton 30284 2003 2004 6 13600 245725.0 -232125.0 Condo NA 92 HAWTHORNE DR Berlin 170028 2017 2017 10 224070 374000.0 -149930.0 null Single Family 395 DAYTON RD Trumbull 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 50320 2005 2006 8 183700 362000.0 -178300.0 Condo NA 159 SCARBOROUGH FARE Berlin 900194 2009 2010 6 0 362031.0 -362031.0 Condo NA 17 IRIS WAY Berlin 170013 2017 2017 10 45150 70000.0 -24850.0 null Condo Family 28A ROBBINS LANE Rocky Hill 100185 2010 2011 6 413300 50000.0 363300.0 Vacant Land NA 501 BERLIN TPKE Berlin 171022 2017 2017 10 31710 80000.0 -48290.0 null Vacant Land 40 AIRPORT RD Killingly 171024 2017 2017 10 93380 140000.0 -46620.0 null Single Family 163 COMMERCE AVE Killingly 170049 2017 2017 10 107500 130000.0 -22500.0 null Single Family 26 FERNDALE DRIVE Manchester 19010 2001 2001 11 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 20076 2002 2003 3 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 170021 2017 2017 10 103140 140000.0 -36860.0 null Single Family 18 HEIDI DR Vernon 50055 2005 2006 2 106580 245000.0 -138420.0 Vacant Lan

#### 3.2. Missing and Null values

In [0]:
from pyspark.sql.functions import col, when, count, isnan

conn3.select(*(count(when(col(x).isNull(), x)).alias(x) 
               for x in conn3.columns)).show()

+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+
SerialNumber|ListYear|SaleYear|SaleMonth|AssessedValue|SaleAmount|PriceDifference|PropertyType|ResidentialType|Address|Town|
+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+
 0| 0| 8| 8| 305| 30793| 31074| 96398| 1| 3| 0|
+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+

Considering we have a dataset of over 912,000 rows, we can safely drop every row containing null values. infact The column with the most missing values, `PropertyType` only has about 10% of it's data missing.

In [0]:
print(conn3.count())

conn32 = conn3.dropna()

# Count after dropping null value rows
print(conn32.count())

912302
784823

In [0]:
conn32.select(*(count(when(col(x).isNull(), x)).alias(x) 
               for x in conn32.columns)).show()

+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+
SerialNumber|ListYear|SaleYear|SaleMonth|AssessedValue|SaleAmount|PriceDifference|PropertyType|ResidentialType|Address|Town|
+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+------------+--------+--------+---------+-------------+----------+---------------+------------+---------------+-------+----+

In [0]:
conn32.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 40043 2004 2005 5 850 50000.0 -49150.0 Vacant Land NA BURNAP BRK RD Andover 30070 2003 2004 8 157300 175000.0 -17700.0 Vacant Land NA ROUTE 6 Andover 40036 2004 2005 4 22400 20000.0 2400.0 Vacant Land NA ROUTE 6 Andover 50249 2005 2006 7 0 197400.0 -197400.0 Condo NA 13 MCINTOSH LN Ansonia 50125 2005 2006 2 331100 375000.0 -43900.0 Condo NA 22 WESTFIELD AVE 5 & Ansonia 50059 2005 2006 5 1196070 3700000.0 -2503930.0 Apartments NA 52 PERRY HL RD Ashford 140042 2014 2015 7 3900 2500.0 1400.0 Vacant Land NA ASHFORD LAKE DR Ashford 900009 2009 2009 12 64000 160000.0 -96000.0 Vacant Land NA HILLSIDE RD Ashford 20384 2002 2003 6 334750 577500.0 -242750.0 Condo NA 1 WILLS WALK Avon 120148 2012 2013 3 125870 125000.0 870.0 Commercial NA 11 CHESTNUT DRIVE Avon 40478 2004 2005 7 109550 229500.0 -119950.0 Condo NA 11 CLOVER CT Avon 20287 2002 2003 4 77000 147500.0 -70500.0 Condo NA 11 WILLOW LN Avon 20345 2002 2003 5 38500 546987.0 -508487.0 Condo NA 12 WYNDEMERE Avon 100064 2010 2010 12 354310 550000.0 -195690.0 Condo NA 16 GATEWOOD Avon 40056 2004 2004 11 105350 175000.0 -69650.0 Condo NA 2 DOVE CIR Avon 60014 2006 2006 10 102900 250000.0 -147100.0 Condo NA 2 SYCAMORE LN Avon 40163 2004 2005 1 307020 430000.0 -122980.0 Condo NA 2 TEMPLETON CT Avon 20194 2002 2003 2 173880 380000.0 -206120.0 Condo NA 3 BRIDLE PATH Avon 10408 2001 2002 7 87080 165000.0 -77920.0 Condo NA 4 SYCAMORE LN Avon 20207 2002 2003 2 125860 235000.0 -109140.0 Condo NA 48 BYRON DR Avon 10340 2001 2002 6 42000 10000.0 32000.0 Vacant Land NA 50 SCHOOL ST Avon 50288 2005 2006 6 170790 339500.0 -168710.0 Condo NA 6 GREENWICH LN Avon 50257 2005 2006 5 153240 284000.0 -130760.0 Condo NA 7 ESSEX CT Avon 20236 2002 2003 3 163910 327500.0 -163590.0 Condo NA 7 MADISON LN Avon 40404 2004 2005 6 189700 350000.0 -160300.0 Condo NA 7 MADISON LN Avon 40384 2004 2005 6 233180 399000.0 -165820.0 Condo NA 70 FOX HOLW Avon 20448 2002 2003 7 87080 189000.0 -101920.0 Condo NA 9 WILLOW LN Avon 30284 2003 2004 6 13600 245725.0 -232125.0 Condo NA 92 HAWTHORNE DR Berlin 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 50320 2005 2006 8 183700 362000.0 -178300.0 Condo NA 159 SCARBOROUGH FARE Berlin 900194 2009 2010 6 0 362031.0 -362031.0 Condo NA 17 IRIS WAY Berlin 100185 2010 2011 6 413300 50000.0 363300.0 Vacant Land NA 501 BERLIN TPKE Berlin 19010 2001 2001 11 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 20076 2002 2003 3 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 50055 2005 2006 2 106580 245000.0 -138420.0 Vacant Land NA 51 JUDD HL RD LT 7 Bethany 10134 2001 2002 6 171350 0.0 171350.0 Residential Single Family 66 TUTTLE CT Bethany 40109 2004 2004 12 121590 244000.0 -122410.0 Condo NA 10 IDLEWOOD Bethel 50053 2005 2005 11 102060 219000.0 -116940.0 Condo NA 20 JUNIPER RD Bethel 60265 2006 2007 8 102060 180000.0 -77940.0 Condo NA 20 JUNIPER RD Bethel 50252 2005 2006 6 123620 280000.0 -156380.0 Condo NA 23 DEER RUN UT 56 Bethel 50212 2005 2006 5 183630 550000.0 -366370.0 Vacant Land NA 28-38 TURKEY PLN RD Bethel 100012 2010 2010 10 142310 160000.0 -17690.0 Condo NA 2B BEACH STREET Bethel 50216 2005 2006 6 128450 260000.0 -131550.0 Condo NA 38 IDLEWOOD UT 122 Bethel 50285 2005 2006 8 184870 328400.0 -143530.0 Condo NA 38 KINGSWOOD DR UT 60 Bethel 30325 2003 2004 6 0 350000.0 -350000.0 Vacant Land NA 7 SARA'S WAY Bethel 50068 2005 2005 12 88200 181000.0 -92800.0 Condo NA 75 JUNIPER RD Bethel 90078 2009 2010 1 124530 140000.0 -15470.0 Condo NA 75 JUNIPER ROAD Bethel 40412 2004 2005 9 67200 96000.0 -28800.0 Commercial NA 7D SCHOOL ST Bethel 100092 2010 2011 2 141050 155000.0 -13950.0 Condo NA 8-O BEACH STREET Bethel 50015 2005 2005 12 83800 180000.0 -96200.0 Vacant Land NA HICKORY LN Bethlehem 50278 2005 2006 4 515250 850000.0 -334750.0 Residential NA 11 MOUNTAIN AVE Bloomfield 50010 2005 2005 10 22765

In [0]:
conn32.describe().display()

summary SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town count 784823 784823 784823 784823 784823 784823 784823 784823 784823 784823 784823 mean 107026.62150828913 2007.4165206677174 2008.183835081286 6.769053659232719 257491.64149750964 358862.7474453348 -101371.10594782517 null null 59443.13333333333 null stddev 525324.1529975114 4.889514060784347 13.668514196929086 3.260254869183219 1301222.6252629762 2038914.3943246175 2115778.8057146627 null null 93004.56597432584 null min 81 2001 2001 1 0 0.0 -9.4094E8 10 Mill Forest Four Family #12 #112 #193 #205 #238 #248 # Andover max 140002850 2016 9999 12 138958820 9.4094E8 1.38918464E8 Vacant Land Two Family parking space only Woodstock

I observed that the `SaleYear` column has a maximum value of **9999**, which is not practical. Let's look at the top values for this column.

In [0]:
conn32.orderBy('SaleYear', ascending=False).display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 60237 2006 9999 1 38500 79900.0 -41400.0 Vacant Land NA 20 MELISSA WAY Stafford 60684 2006 9999 1 0 83400.0 -83400.0 Residential Single Family 70 N TURNPIKE RD Wallingford 110276 2011 2102 8 208800 265000.0 -56200.0 Residential Single Family 5 ROSEMARY LN Guilford 160412 2016 2017 5 64740 38410.0 26330.0 Residential Single Family 476 HIGH ST Torrington 160499 2016 2017 6 124300 168000.0 -43700.0 Residential Single Family 71 CANTERBURY TPKE Norwich 161477 2016 2017 7 998180 1350000.0 -351820.0 Residential Single Family 2 WALLENBERG DRIVE Stamford 161780 2016 2017 8 171830 244000.0 -72170.0 Residential Single Family 127 GREYROCK PLACE # 1005 Stamford 161471 2016 2017 7 218810 370000.0 -151190.0 Residential Single Family 850 EAST MAIN STREET # 418 Stamford 160718 2016 2017 5 117670 217000.0 -99330.0 Residential Single Family 255 HEPBURN RD Hamden 160249 2016 2017 9 205050 290000.0 -84950.0 Residential Single Family 41 BENNETT CIRCLE Voluntown 161081 2016 2017 7 91350 137000.0 -45650.0 Residential Single Family 32 PEARL LAKE RD Waterbury 160276 2016 2017 6 217280 352500.0 -135220.0 Residential Two Family 136B & G34 HERITAGE HILL RD New Canaan 161096 2016 2017 9 134750 90000.0 44750.0 Residential Single Family 2065 SHEPARD AVE Hamden 160220 2016 2017 3 206220 275000.0 -68780.0 Residential Single Family 18 WHITE OAK RD Trumbull 160015 2016 2017 4 126840 229900.0 -103060.0 Residential Single Family 93 SCHOOL HILL RD Sprague 160591 2016 2017 5 1294500 2600000.0 -1305500.0 Industrial NA 63-79 TRIANGLE ST/LEE MAC AVE Danbury 160121 2016 2017 1 1576200 2958000.0 -1381800.0 Residential Single Family 11 ELMWOOD RD Westport 160123 2016 2017 4 214500 358000.0 -143500.0 Residential Single Family 82 GILLOTTI RD New Fairfield 16089 2016 2017 3 108580 165000.0 -56420.0 Vacant Land NA 89 WEST LEDGE Burlington 160148 2016 2017 5 158100 325000.0 -166900.0 Residential Single Family 505 FIVE MILE RIVER RD Putnam 160375 2016 2017 3 107240 83000.0 24240.0 Residential Single Family 116 COTTAGE ST Meriden 160372 2016 2017 1 119630 212000.0 -92370.0 Residential Single Family 40 ROCKWELL AVE Stratford 160730 2016 2017 6 35000 730589.0 -695589.0 Residential Single Family 73 WINDING RIDGE WAY Danbury 160794 2016 2017 7 854000 1150000.0 -296000.0 Residential Single Family 1005 BURR STREET Fairfield 160815 2016 2017 7 509180 650000.0 -140820.0 Residential Single Family 193 HAMILTON AVE 18 Greenwich 160112 2016 2017 4 172670 254025.0 -81355.0 Residential Single Family 38 CINDY RD Ellington 160165 2016 2017 1 101100 133775.0 -32675.0 Residential Single Family 8P ST MARC CIRCLE South Windsor 160064 2016 2017 1 207100 363785.0 -156685.0 Residential Single Family 711 CHAMPIONSHIP DR Oxford 160164 2016 2017 7 330100 482450.0 -152350.0 Residential Single Family 10 OWL RIDGE RD Oxford 160940 2016 2017 6 55840 50705.0 5135.0 Residential Single Family 101 SHEFFIELD ST Waterbury 160208 2016 2017 8 162270 202000.0 -39730.0 Residential Single Family 705 MEADOWVIEW DR East Windsor 160160 2016 2017 4 122150 156000.0 -33850.0 Residential Single Family 4 TANAGER LN Ledyard 160300 2016 2017 2 52903760 470000.0 5.243376E7 Commercial NA 196 PUTNAM PARK Greenwich 160467 2016 2017 7 248630 424900.0 -176270.0 Residential Single Family 126 CHARTER OAK RD Southbury 160498 2016 2017 8 345730 535000.0 -189270.0 Residential Single Family 75 GATEHOUSE RD Trumbull 16268 2016 2017 8 389100 575000.0 -185900.0 Residential Single Family 276 HAMMERTOWN RD Monroe 161283 2016 2017 6 289730 484900.0 -195170.0 Residential Single Family 175 WEST AVENUE # 7 Stamford 160391 2016 2017 7 107450 149900.0 -42450.0 Residential Single Family 25 EARLY DAWN CIR Windsor 160322 2016 2017 5 166400 290000.0 -123600.0 Residential Single Family 91 LEFOLL BLVD South Windsor 160555 2016 2017 4 190470 260000.0 -69530.0 Residential Single Family 170 HILLFIELD RD Hamden 160135 2016 2017 7 150400 26

We can thus drop the first three rows since their Sale Years are wrong. The dataset contains records for only houses sold between 2001 and 2018

In [0]:
conn32 =conn32.filter(conn32.SaleYear<2018)
conn32.select('SaleYear').describe().display()

summary SaleYear count 784820 mean 2008.1633521062154 stddev 4.908836362606605 min 2001 max 2017

## 4  Visualization and Analysis

#### 4.1. Grouping by listing and sale years

In [0]:
# Rename dataframe to be more generic
df = conn32
# How did total sale volume fare over the years
conn411 = df.groupBy('SaleYear').agg({'SaleAmount':'sum','SerialNumber':'count'}).display()

SaleYear count(SerialNumber) sum(SaleAmount) 2003 84802 2.5603846106E10 2007 44058 2.08327185634E10 2015 47153 1.791333567467E10 2006 54818 2.0293195147E10 2013 36029 1.4117630457E10 2014 40314 1.5401226661E10 2004 71042 2.3526528692E10 2012 31239 1.2157705140220001E10 2009 34833 1.1068288456E10 2016 45919 2.1438690332E10 2001 13854 3.019735355E9 2005 76473 2.8518873803E10 2010 37464 1.2968570129E10 2011 25356 8.308172256E9 2008 32715 1.294728217E10 2017 35916 1.3793701932E10 2002 72835 1.9733808864E10

In [0]:
# How did the assessed values compare with actual values over the years

conn412 = df.groupBy('SaleYear').agg({'SaleAmount':'avg','AssessedValue':'avg'}).display()

SaleYear avg(SaleAmount) avg(AssessedValue) 2003 301925.02660314617 178091.03214546826 2007 472847.57736165967 343262.9041944709 2015 379898.11199011724 333951.69363561174 2006 370192.18408187095 215733.88270276185 2013 391840.7520885953 341099.1027505621 2014 382031.717542293 314406.3624051198 2004 331163.6594127418 191733.59559134033 2012 389183.5570991389 398016.7310093153 2009 317752.9485258232 287815.62633135245 2016 466880.60131971515 264717.04242252663 2001 217968.48238775803 132143.82683701458 2005 372927.35740718944 205158.41806912245 2010 346160.8511904762 338133.47918001283 2011 327660.9976336962 314052.1959299574 2008 395759.8095674767 317304.4784655357 2017 384054.5141997995 317866.91713999334 2002 270938.5441614608 152922.76623875884

In [0]:
# Let's see how the top three have differed over the years
conn422 = df.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 40043 2004 2005 5 850 50000.0 -49150.0 Vacant Land NA BURNAP BRK RD Andover 30070 2003 2004 8 157300 175000.0 -17700.0 Vacant Land NA ROUTE 6 Andover 40036 2004 2005 4 22400 20000.0 2400.0 Vacant Land NA ROUTE 6 Andover 50249 2005 2006 7 0 197400.0 -197400.0 Condo NA 13 MCINTOSH LN Ansonia 50125 2005 2006 2 331100 375000.0 -43900.0 Condo NA 22 WESTFIELD AVE 5 & Ansonia 50059 2005 2006 5 1196070 3700000.0 -2503930.0 Apartments NA 52 PERRY HL RD Ashford 140042 2014 2015 7 3900 2500.0 1400.0 Vacant Land NA ASHFORD LAKE DR Ashford 900009 2009 2009 12 64000 160000.0 -96000.0 Vacant Land NA HILLSIDE RD Ashford 20384 2002 2003 6 334750 577500.0 -242750.0 Condo NA 1 WILLS WALK Avon 120148 2012 2013 3 125870 125000.0 870.0 Commercial NA 11 CHESTNUT DRIVE Avon 40478 2004 2005 7 109550 229500.0 -119950.0 Condo NA 11 CLOVER CT Avon 20287 2002 2003 4 77000 147500.0 -70500.0 Condo NA 11 WILLOW LN Avon 20345 2002 2003 5 38500 546987.0 -508487.0 Condo NA 12 WYNDEMERE Avon 100064 2010 2010 12 354310 550000.0 -195690.0 Condo NA 16 GATEWOOD Avon 40056 2004 2004 11 105350 175000.0 -69650.0 Condo NA 2 DOVE CIR Avon 60014 2006 2006 10 102900 250000.0 -147100.0 Condo NA 2 SYCAMORE LN Avon 40163 2004 2005 1 307020 430000.0 -122980.0 Condo NA 2 TEMPLETON CT Avon 20194 2002 2003 2 173880 380000.0 -206120.0 Condo NA 3 BRIDLE PATH Avon 10408 2001 2002 7 87080 165000.0 -77920.0 Condo NA 4 SYCAMORE LN Avon 20207 2002 2003 2 125860 235000.0 -109140.0 Condo NA 48 BYRON DR Avon 10340 2001 2002 6 42000 10000.0 32000.0 Vacant Land NA 50 SCHOOL ST Avon 50288 2005 2006 6 170790 339500.0 -168710.0 Condo NA 6 GREENWICH LN Avon 50257 2005 2006 5 153240 284000.0 -130760.0 Condo NA 7 ESSEX CT Avon 20236 2002 2003 3 163910 327500.0 -163590.0 Condo NA 7 MADISON LN Avon 40404 2004 2005 6 189700 350000.0 -160300.0 Condo NA 7 MADISON LN Avon 40384 2004 2005 6 233180 399000.0 -165820.0 Condo NA 70 FOX HOLW Avon 20448 2002 2003 7 87080 189000.0 -101920.0 Condo NA 9 WILLOW LN Avon 30284 2003 2004 6 13600 245725.0 -232125.0 Condo NA 92 HAWTHORNE DR Berlin 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 50320 2005 2006 8 183700 362000.0 -178300.0 Condo NA 159 SCARBOROUGH FARE Berlin 900194 2009 2010 6 0 362031.0 -362031.0 Condo NA 17 IRIS WAY Berlin 100185 2010 2011 6 413300 50000.0 363300.0 Vacant Land NA 501 BERLIN TPKE Berlin 19010 2001 2001 11 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 20076 2002 2003 3 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 50055 2005 2006 2 106580 245000.0 -138420.0 Vacant Land NA 51 JUDD HL RD LT 7 Bethany 10134 2001 2002 6 171350 0.0 171350.0 Residential Single Family 66 TUTTLE CT Bethany 40109 2004 2004 12 121590 244000.0 -122410.0 Condo NA 10 IDLEWOOD Bethel 50053 2005 2005 11 102060 219000.0 -116940.0 Condo NA 20 JUNIPER RD Bethel 60265 2006 2007 8 102060 180000.0 -77940.0 Condo NA 20 JUNIPER RD Bethel 50252 2005 2006 6 123620 280000.0 -156380.0 Condo NA 23 DEER RUN UT 56 Bethel 50212 2005 2006 5 183630 550000.0 -366370.0 Vacant Land NA 28-38 TURKEY PLN RD Bethel 100012 2010 2010 10 142310 160000.0 -17690.0 Condo NA 2B BEACH STREET Bethel 50216 2005 2006 6 128450 260000.0 -131550.0 Condo NA 38 IDLEWOOD UT 122 Bethel 50285 2005 2006 8 184870 328400.0 -143530.0 Condo NA 38 KINGSWOOD DR UT 60 Bethel 30325 2003 2004 6 0 350000.0 -350000.0 Vacant Land NA 7 SARA'S WAY Bethel 50068 2005 2005 12 88200 181000.0 -92800.0 Condo NA 75 JUNIPER RD Bethel 90078 2009 2010 1 124530 140000.0 -15470.0 Condo NA 75 JUNIPER ROAD Bethel 40412 2004 2005 9 67200 96000.0 -28800.0 Commercial NA 7D SCHOOL ST Bethel 100092 2010 2011 2 141050 155000.0 -13950.0 Condo NA 8-O BEACH STREET Bethel 50015 2005 2005 12 83800 180000.0 -96200.0 Vacant Land NA HICKORY LN Bethlehem 50278 2005 2006 4 515250 850000.0 -334750.0 Residential NA 11 MOUNTAIN AVE Bloomfield 50010 2005 2005 10 22765

#### 4.2. Group by property type

In [0]:
# What are the average sale prices for the different property types
conn421 = df.groupBy('PropertyType').agg({'SaleAmount':'avg'}).display()

PropertyType avg(SaleAmount) Apartments 1198831.0944005742 Vacant Land 198013.20119754985 NA 680015.4588020674 Residential 351985.14680104627 Industrial 997839.3087719298 Condo 214239.4411592828 Public Utility 222799.6015037594 10 Mill Forest 183016.0 Commercial 1040685.0529760687

In [0]:
# On average, how accurate were the valuations compared to the eventual sale prices for different property types.
conn423 = df.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 40043 2004 2005 5 850 50000.0 -49150.0 Vacant Land NA BURNAP BRK RD Andover 30070 2003 2004 8 157300 175000.0 -17700.0 Vacant Land NA ROUTE 6 Andover 40036 2004 2005 4 22400 20000.0 2400.0 Vacant Land NA ROUTE 6 Andover 50249 2005 2006 7 0 197400.0 -197400.0 Condo NA 13 MCINTOSH LN Ansonia 50125 2005 2006 2 331100 375000.0 -43900.0 Condo NA 22 WESTFIELD AVE 5 & Ansonia 50059 2005 2006 5 1196070 3700000.0 -2503930.0 Apartments NA 52 PERRY HL RD Ashford 140042 2014 2015 7 3900 2500.0 1400.0 Vacant Land NA ASHFORD LAKE DR Ashford 900009 2009 2009 12 64000 160000.0 -96000.0 Vacant Land NA HILLSIDE RD Ashford 20384 2002 2003 6 334750 577500.0 -242750.0 Condo NA 1 WILLS WALK Avon 120148 2012 2013 3 125870 125000.0 870.0 Commercial NA 11 CHESTNUT DRIVE Avon 40478 2004 2005 7 109550 229500.0 -119950.0 Condo NA 11 CLOVER CT Avon 20287 2002 2003 4 77000 147500.0 -70500.0 Condo NA 11 WILLOW LN Avon 20345 2002 2003 5 38500 546987.0 -508487.0 Condo NA 12 WYNDEMERE Avon 100064 2010 2010 12 354310 550000.0 -195690.0 Condo NA 16 GATEWOOD Avon 40056 2004 2004 11 105350 175000.0 -69650.0 Condo NA 2 DOVE CIR Avon 60014 2006 2006 10 102900 250000.0 -147100.0 Condo NA 2 SYCAMORE LN Avon 40163 2004 2005 1 307020 430000.0 -122980.0 Condo NA 2 TEMPLETON CT Avon 20194 2002 2003 2 173880 380000.0 -206120.0 Condo NA 3 BRIDLE PATH Avon 10408 2001 2002 7 87080 165000.0 -77920.0 Condo NA 4 SYCAMORE LN Avon 20207 2002 2003 2 125860 235000.0 -109140.0 Condo NA 48 BYRON DR Avon 10340 2001 2002 6 42000 10000.0 32000.0 Vacant Land NA 50 SCHOOL ST Avon 50288 2005 2006 6 170790 339500.0 -168710.0 Condo NA 6 GREENWICH LN Avon 50257 2005 2006 5 153240 284000.0 -130760.0 Condo NA 7 ESSEX CT Avon 20236 2002 2003 3 163910 327500.0 -163590.0 Condo NA 7 MADISON LN Avon 40404 2004 2005 6 189700 350000.0 -160300.0 Condo NA 7 MADISON LN Avon 40384 2004 2005 6 233180 399000.0 -165820.0 Condo NA 70 FOX HOLW Avon 20448 2002 2003 7 87080 189000.0 -101920.0 Condo NA 9 WILLOW LN Avon 30284 2003 2004 6 13600 245725.0 -232125.0 Condo NA 92 HAWTHORNE DR Berlin 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 50320 2005 2006 8 183700 362000.0 -178300.0 Condo NA 159 SCARBOROUGH FARE Berlin 900194 2009 2010 6 0 362031.0 -362031.0 Condo NA 17 IRIS WAY Berlin 100185 2010 2011 6 413300 50000.0 363300.0 Vacant Land NA 501 BERLIN TPKE Berlin 19010 2001 2001 11 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 20076 2002 2003 3 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 50055 2005 2006 2 106580 245000.0 -138420.0 Vacant Land NA 51 JUDD HL RD LT 7 Bethany 10134 2001 2002 6 171350 0.0 171350.0 Residential Single Family 66 TUTTLE CT Bethany 40109 2004 2004 12 121590 244000.0 -122410.0 Condo NA 10 IDLEWOOD Bethel 50053 2005 2005 11 102060 219000.0 -116940.0 Condo NA 20 JUNIPER RD Bethel 60265 2006 2007 8 102060 180000.0 -77940.0 Condo NA 20 JUNIPER RD Bethel 50252 2005 2006 6 123620 280000.0 -156380.0 Condo NA 23 DEER RUN UT 56 Bethel 50212 2005 2006 5 183630 550000.0 -366370.0 Vacant Land NA 28-38 TURKEY PLN RD Bethel 100012 2010 2010 10 142310 160000.0 -17690.0 Condo NA 2B BEACH STREET Bethel 50216 2005 2006 6 128450 260000.0 -131550.0 Condo NA 38 IDLEWOOD UT 122 Bethel 50285 2005 2006 8 184870 328400.0 -143530.0 Condo NA 38 KINGSWOOD DR UT 60 Bethel 30325 2003 2004 6 0 350000.0 -350000.0 Vacant Land NA 7 SARA'S WAY Bethel 50068 2005 2005 12 88200 181000.0 -92800.0 Condo NA 75 JUNIPER RD Bethel 90078 2009 2010 1 124530 140000.0 -15470.0 Condo NA 75 JUNIPER ROAD Bethel 40412 2004 2005 9 67200 96000.0 -28800.0 Commercial NA 7D SCHOOL ST Bethel 100092 2010 2011 2 141050 155000.0 -13950.0 Condo NA 8-O BEACH STREET Bethel 50015 2005 2005 12 83800 180000.0 -96200.0 Vacant Land NA HICKORY LN Bethlehem 50278 2005 2006 4 515250 850000.0 -334750.0 Residential NA 11 MOUNTAIN AVE Bloomfield 50010 2005 2005 10 22765

#### 4.3. Group by Town

In [0]:
conn431 = df.groupby('Town').agg({'SaleAmount':'sum'}).orderBy('sum(SaleAmount)', ascending=False)
conn431.display()
# We can see the top 10 towns by total sales over the years

Town sum(SaleAmount) Greenwich 2.5321389878E10 Stamford 2.018732831E10 Westport 1.0012903568E10 Norwalk 9.735927797E9 Fairfield 8.659282672E9 Darien 7.720055425E9 New Canaan 7.130430456E9 Danbury 5.625856634E9 Bridgeport 5.455730714E9 West Hartford 5.241808613E9 Milford 4.335240936E9 Ridgefield 4.134757486E9 New Haven 3.942723258E9 Hartford 3.665081761E9 Rocky Hill 3.6521092024E9 Glastonbury 3.571976699E9 Wilton 3.47785352E9 Hamden 3.331993357E9 Shelton 3.296899055E9 Trumbull 3.26315839367E9 Stratford 3.206182167E9 Waterbury 3.177239218E9 Manchester 3.008630718E9 Wallingford 2.78173463E9 Farmington 2.699462897E9 Newtown 2.691886454E9 Avon 2.677387593E9 Southington 2.56279485E9 Meriden 2.523499177E9 Bristol 2.475298078E9 Weston 2.379773005E9 Middletown 2.263437826E9 Guilford 2.203723579E9 Madison 2.199120994E9 Windsor 2.191118831E9 New Britain 2.135634128E9 New Milford 2.113286895E9 Cheshire 2.092857902E9 Simsbury 2.021564827E9 West Haven 2.019171113E9 Groton 2.005010871E9 Newington 1.969790579E9 East Hartford 1.858397063E9 Branford 1.830862675E9 Stonington 1.824015019E9 South Windsor 1.77304639E9 Monroe 1.765661233E9 Enfield 1.756464362E9 Brookfield 1.720577477E9 Torrington 1.712233719E9 North Haven 1.698245707E9 Southbury 1.694863591E9 Norwich 1.631818648E9 East Lyme 1.516988768E9 Wethersfield 1.505264174E9 East Haven 1.496408748E9 Orange 1.449742892E9 Bloomfield 1.405723963E9 Bethel 1.37357268E9 Old Saybrook 1.371674849E9 Vernon 1.34687367E9 Waterford 1.256222006E9 Naugatuck 1.249616766E9 New Fairfield 1.208867212E9 Berlin 1.198818688E9 Oxford 1.142416797E9 Redding 1.11293972E9 Suffield 1.085456668E9 Watertown 1.083377632E9 Washington 1.034626173E9 Cromwell 1.026104762E9 Clinton 1.021615496E9 New London 1.008117702E9 Ellington 9.60923282E8 Easton 9.48780531E8 Canton 9.35656589E8 Old Lyme 8.95854475E8 Essex 8.83667489E8 East Hampton 8.78650717E8 Tolland 8.74239626E8 Montville 8.62857436E8 Woodbridge 8.62296972E8 Plainville 8.44616755E8 Woodbury 8.42683393E8 Seymour 8.24462792E8 Colchester 8.02490039E8 Mansfield 7.91571456E8 Killingly 7.88158005E8 Windsor Locks 7.85254513E8 Ledyard 7.73897362E8 Wolcott 7.66889549E8 Granby 7.60354061E8 Burlington 7.02494633E8 Litchfield 6.84671952E8 North Branford 6.61536599E8 East Windsor 6.59986348E8 Plainfield 6.55464325E8 Westbrook 6.45331161E8 Windham 6.44567322E8 Coventry 6.42812203E8 Salisbury 6.25561416E8 Ansonia 6.23339004E8 Somers 5.80168159E8 East Haddam 5.77410759E8 Portland 5.68414287E8 Middlebury 5.57789462E8 Prospect 5.31330844E8 Hebron 5.28440986E8 Haddam 5.04361697E8 Derby 4.94725418E8 Stafford 4.8789993E8 Winchester 4.84720314E8 Woodstock 4.80049507E8 Sherman 4.75106984E8 Griswold 4.5134959922E8 Killingworth 4.48780273E8 Durham 4.44330512E8 Plymouth 4.33585383E8 Sharon 4.30756018E8 Roxbury 4.29912383E8 Putnam 4.01725777E8 New Hartford 4.00732329E8 Kent 3.94839658E8 East Granby 3.92024437E8 Marlborough 3.86806132E8 Brooklyn 3.76253681E8 Thompson 3.75599659E8 Bethany 3.73618018E8 Goshen 3.68014872E8 Lebanon 3.59425238E8 Beacon Falls 3.57100433E8 Thomaston 3.43423242E8 North Stonington 3.29503635E8 Lyme 3.00405454E8 Chester 2.87222466E8 Columbia 2.76804777E8 Bolton 2.74433421E8 Deep River 2.70161068E8 Harwinton 2.70052445E8 Salem 2.53807456E8 Willington 2.36756338E8 Warren 2.33245103E8 Preston 2.15875625E8 Pomfret 2.11495837E8 Bridgewater 2.06464545E8 Canterbury 1.91654534E8 Ashford 1.89822397E8 Middlefield 1.8728164E8 Sterling 1.7767161E8 Bethlehem 1.77144393E8 Morris 1.76857637E8 Barkhamsted 1.64340819E8 Lisbon 1.48800999E8 Norfolk 1.4657264E8 Cornwall 1.40560153E8 Andover 1.35634814E8 North Canaan 1.14394224E8 Voluntown 1.10156042E8 Sprague 9.8472554E7 Hampton 8.5553004E7 Franklin 8.5456945E7 Bozrah 8.2721153E7 Colebrook 8.1095982E7 Eastford 7.7189185E7 Canaan 7.5889122E7 Chaplin 7.4555213E7 Hartland 6.3974249E7 Scotland 6.0456468E7 Union 4.1330478E7

In [0]:
conn432 = df.groupby('Town').agg({'SaleAmount':'avg'}).orderBy('avg(SaleAmount)', ascending=False)
conn432.display()

Town avg(SaleAmount) Greenwich 1977615.579350203 New Canaan 1480571.107973422 Darien 1478937.820881226 Westport 1337729.2676018705 Weston 942484.3584158416 Wilton 897279.0299277606 Washington 835049.3728813559 Stamford 771893.4084043895 Rocky Hill 744568.6447298675 Ridgefield 740730.4704407023 Roxbury 702471.2140522876 Fairfield 679052.9071518193 Easton 634211.5848930482 Redding 629134.9462973431 Bridgewater 521375.11363636365 Madison 520502.0104142012 Salisbury 520001.17705735663 Orange 513365.04674220964 Norwalk 511125.9868227635 Warren 504859.5303030303 Lyme 476078.3740095087 Avon 454179.4050890585 Woodbridge 454079.5007898894 Newtown 452570.01580363145 Trumbull 452274.2056368677 Sherman 444856.7265917603 Guilford 428573.23590042786 Essex 426686.3780782231 Sharon 419840.17348927876 Farmington 406178.58817333734 Brookfield 404366.03454759106 Old Saybrook 394613.0175489068 Glastonbury 392697.52627528587 Kent 392484.74950298213 Stonington 387675.8807651435 Monroe 383589.2315880947 Middlebury 374104.266934943 Shelton 369690.40760260145 Danbury 362958.49251612904 New Fairfield 361827.9592936247 Old Lyme 361523.1941081517 Simsbury 359645.0501690091 North Haven 342664.5897901534 Litchfield 340632.81194029853 West Hartford 338203.0203884122 Oxford 337294.5961027458 Bethel 337238.56616744417 Cheshire 333735.911656833 Westbrook 333504.4759689922 Cornwall 331509.7948113208 Chester 330140.7655172414 Goshen 317527.93097497843 Milford 316926.74435265735 Killingworth 316265.1677237491 East Lyme 313362.68704813055 Windsor 308347.7105263158 Woodbury 308336.4043175997 Canton 301630.1060606061 Suffield 301097.55006934813 New Milford 298740.01908396947 Burlington 294917.9819479429 Bloomfield 290618.9710564399 Branford 290474.80168173887 Southbury 290017.7260438056 Berlin 288663.30074644834 Durham 287035.21447028423 Norfolk 284055.503875969 Granby 283926.0870052278 Morris 283425.7003205128 Somers 281635.0286407767 South Windsor 281614.7379288437 Wallingford 278786.7939466827 Marlborough 278478.1367890569 Bethany 275936.4977843427 East Granby 273569.04187020235 Groton 272864.84363092 Bethlehem 266783.7243975904 Clinton 266115.0028653295 Prospect 265532.65567216394 Tolland 265001.40224310395 Deep River 264345.4677103718 North Branford 262723.03375694994 Haddam 262142.25415800416 Southington 261482.99663299663 Waterford 258428.71960501955 Beacon Falls 257462.46070656093 Cromwell 255950.3023197805 Hebron 253327.414189837 Ellington 251880.28361730013 Canaan 249635.2697368421 Newington 248993.87928201238 Bolton 247237.31621621622 Wethersfield 246401.07611720412 Hamden 241466.29154286542 Mansfield 241185.69652650823 Stratford 240361.50888372443 Middletown 240280.02399150742 Middlefield 239797.2343149808 Hartford 239141.44336421767 East Haddam 238107.5294845361 North Stonington 238080.66112716764 Harwinton 235237.32142857142 Watertown 234751.38288190681 Manchester 233680.0557669903 Willington 231886.71694417237 Colchester 231131.92367511522 Seymour 230748.05261684858 East Windsor 229720.27427775844 Windsor Locks 226428.63696655133 Pomfret 225715.94130202776 Portland 225382.3501189532 New Haven 222012.68416014416 New Hartford 222012.3706371191 Columbia 221621.11849479584 Wolcott 219739.1257879656 Enfield 216260.0790445703 East Hampton 216150.23788437885 East Haven 215776.31550108147 Ledyard 211505.1549603717 Barkhamsted 208290.0114068441 Salem 207698.4091653028 Franklin 205920.34939759035 Andover 204886.42598187312 Montville 204662.57969639468 Bozrah 204249.76049382717 Colebrook 201230.72456575683 Vernon 198390.58329650905 Ansonia 197196.77443846883 New London 197090.45982404693 West Haven 194281.83517752335 Coventry 194085.81008454107 Hartland 193861.3606060606 Bristol 191883.5719379845 Union 191344.80555555556 Eastford 191062.33910891088 Ashford 190968.20623742454 Brooklyn 188503.8481963928 Woodstock 188476.44562230076 Plainfield 188351.81752873564 Plainville 187943.20315976857 Preston 187229.50997398092 Derby 185777.4757791964 Lebanon 185366.29087158

#### 4.4.  Analysis: The 2008 Real Estate Crash

We can inspect what property type dominated sales by quantity.

In [0]:
conn441 = df.filter('SaleYear>2007 and SaleYear<2009')
conn441.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 70019 2007 2008 4 198100 350000.0 -151900.0 Residential Single Family 111 LAKESIDE DR Andover 70022 2007 2008 6 102300 172500.0 -70200.0 Residential Single Family 12 ROSE LN Andover 70037 2007 2008 7 178400 239000.0 -60600.0 Residential Single Family 146 LAKESIDE DR Andover 70027 2007 2008 8 260200 250000.0 10200.0 Residential Single Family 15 DOGWOOD AVENUE Andover 70023 2007 2008 7 215800 325000.0 -109200.0 Residential Single Family 177 SHODDY MILL RD Andover 70036 2007 2008 8 154200 224900.0 -70700.0 Residential Single Family 18 LAKESIDE DR Andover 80007 2008 2008 12 155600 210000.0 -54400.0 Residential Single Family 19 PARKER BRIDGE ROAD Andover 70026 2007 2008 8 131500 221000.0 -89500.0 Residential Single Family 2 MERRITT VALLEY ROAD Andover 80001 2008 2008 10 91400 90000.0 1400.0 Residential Single Family 233 LAKE ROAD Andover 70020 2007 2008 5 246700 365000.0 -118300.0 Residential Single Family 242 GILEAD RD Andover 70029 2007 2008 8 109000 170000.0 -61000.0 Residential Single Family 25 ASPINALL DRIVE Andover 70040 2007 2008 6 291400 407000.0 -115600.0 Residential Single Family 29 STANLEY DR Andover 70016 2007 2008 3 114900 179770.0 -64870.0 Residential Single Family 30 ROUTE 6 Andover 80002 2008 2008 11 114900 102000.0 12900.0 Residential Single Family 30 ROUTE 6 Andover 70033 2007 2008 6 118600 180000.0 -61400.0 Residential Single Family 321 LAKE RD Andover 80006 2008 2008 12 124900 195500.0 -70600.0 Residential Single Family 327 LAKE ROAD Andover 70028 2007 2008 8 112500 175000.0 -62500.0 Residential Single Family 33 GILEAD ROAD Andover 70018 2007 2008 3 312900 382500.0 -69600.0 Residential Single Family 336 HEBRON RD Andover 70039 2007 2008 6 181400 285000.0 -103600.0 Residential Single Family 343 LAKE RD Andover 70025 2007 2008 7 327510 595000.0 -267490.0 Residential Single Family 36 OAK FARMS RD Andover 80005 2008 2008 11 186400 150000.0 36400.0 Residential Single Family 371 LAKE ROAD & LOT 5 LAKE ROA Andover 70017 2007 2008 3 79100 391500.0 -312400.0 Residential Single Family 4 SHADBLOW LN Andover 70034 2007 2008 9 158600 260000.0 -101400.0 Residential Single Family 44 WALES RD Andover 70038 2007 2008 9 304400 420000.0 -115600.0 Residential Single Family 5 SHADBLOW LN Andover 70013 2007 2008 2 214600 231000.0 -16400.0 Residential Single Family 51 WINDRUSH LN Andover 70014 2007 2008 1 147200 170000.0 -22800.0 Residential Single Family 587 ROUTE 6 Andover 80003 2008 2008 11 175100 250000.0 -74900.0 Residential Single Family 61 SHODDY MILL ROAD Andover 70035 2007 2008 9 133100 213750.0 -80650.0 Residential Single Family 76 WEST ST Andover 70021 2007 2008 6 169200 306200.0 -137000.0 Residential Single Family 85 GILEAD RD Andover 70031 2007 2008 9 209200 270600.0 -61400.0 Residential Single Family 901 EAST STREET Andover 80004 2008 2008 11 152700 175000.0 -22300.0 Residential Single Family 92 HUTCHINSON Andover 70024 2007 2008 7 188600 260000.0 -71400.0 Residential Single Family 93 LAKESIDE DR Andover 80013 2008 2008 10 10500 40250.0 -29750.0 Vacant Land NA LOT 12-1 BUNKER HILL RD Andover 800063 2008 2008 12 148300 149900.0 -1600.0 Residential Two Family 114-116 MYRTLE AVE Ansonia 80013 2008 2008 10 146000 17000.0 129000.0 Residential Two Family 114 TREMONT ST Ansonia 800061 2008 2008 12 143500 90000.0 53500.0 Residential Two Family 13-15 EAGLE ST Ansonia 80006 2008 2008 10 131700 188000.0 -56300.0 Residential Single Family 13 HIGHLAND TERR Ansonia 800051 2008 2008 12 173400 265000.0 -91600.0 Residential Single Family 136 BENZ ST Ansonia 800077 2008 2008 12 173400 236000.0 -62600.0 Residential Single Family 136 BENZ ST Ansonia 800086 2008 2008 11 164600 250000.0 -85400.0 Residential Single Family 136 HILL ST Ansonia 80016 2008 2008 10 205000 205000.0 0.0 Residential Three Family 139-141 N STATE ST Ansonia 80003 2008 2008

With residential properties accounting for the highest propertion of sale volume in 2008, we take a closer look at what types of residential homes were most prominent.

In [0]:
conn442 = df.filter('SaleYear>2007 and SaleYear<2009')
conn442.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 70019 2007 2008 4 198100 350000.0 -151900.0 Residential Single Family 111 LAKESIDE DR Andover 70022 2007 2008 6 102300 172500.0 -70200.0 Residential Single Family 12 ROSE LN Andover 70037 2007 2008 7 178400 239000.0 -60600.0 Residential Single Family 146 LAKESIDE DR Andover 70027 2007 2008 8 260200 250000.0 10200.0 Residential Single Family 15 DOGWOOD AVENUE Andover 70023 2007 2008 7 215800 325000.0 -109200.0 Residential Single Family 177 SHODDY MILL RD Andover 70036 2007 2008 8 154200 224900.0 -70700.0 Residential Single Family 18 LAKESIDE DR Andover 80007 2008 2008 12 155600 210000.0 -54400.0 Residential Single Family 19 PARKER BRIDGE ROAD Andover 70026 2007 2008 8 131500 221000.0 -89500.0 Residential Single Family 2 MERRITT VALLEY ROAD Andover 80001 2008 2008 10 91400 90000.0 1400.0 Residential Single Family 233 LAKE ROAD Andover 70020 2007 2008 5 246700 365000.0 -118300.0 Residential Single Family 242 GILEAD RD Andover 70029 2007 2008 8 109000 170000.0 -61000.0 Residential Single Family 25 ASPINALL DRIVE Andover 70040 2007 2008 6 291400 407000.0 -115600.0 Residential Single Family 29 STANLEY DR Andover 70016 2007 2008 3 114900 179770.0 -64870.0 Residential Single Family 30 ROUTE 6 Andover 80002 2008 2008 11 114900 102000.0 12900.0 Residential Single Family 30 ROUTE 6 Andover 70033 2007 2008 6 118600 180000.0 -61400.0 Residential Single Family 321 LAKE RD Andover 80006 2008 2008 12 124900 195500.0 -70600.0 Residential Single Family 327 LAKE ROAD Andover 70028 2007 2008 8 112500 175000.0 -62500.0 Residential Single Family 33 GILEAD ROAD Andover 70018 2007 2008 3 312900 382500.0 -69600.0 Residential Single Family 336 HEBRON RD Andover 70039 2007 2008 6 181400 285000.0 -103600.0 Residential Single Family 343 LAKE RD Andover 70025 2007 2008 7 327510 595000.0 -267490.0 Residential Single Family 36 OAK FARMS RD Andover 80005 2008 2008 11 186400 150000.0 36400.0 Residential Single Family 371 LAKE ROAD & LOT 5 LAKE ROA Andover 70017 2007 2008 3 79100 391500.0 -312400.0 Residential Single Family 4 SHADBLOW LN Andover 70034 2007 2008 9 158600 260000.0 -101400.0 Residential Single Family 44 WALES RD Andover 70038 2007 2008 9 304400 420000.0 -115600.0 Residential Single Family 5 SHADBLOW LN Andover 70013 2007 2008 2 214600 231000.0 -16400.0 Residential Single Family 51 WINDRUSH LN Andover 70014 2007 2008 1 147200 170000.0 -22800.0 Residential Single Family 587 ROUTE 6 Andover 80003 2008 2008 11 175100 250000.0 -74900.0 Residential Single Family 61 SHODDY MILL ROAD Andover 70035 2007 2008 9 133100 213750.0 -80650.0 Residential Single Family 76 WEST ST Andover 70021 2007 2008 6 169200 306200.0 -137000.0 Residential Single Family 85 GILEAD RD Andover 70031 2007 2008 9 209200 270600.0 -61400.0 Residential Single Family 901 EAST STREET Andover 80004 2008 2008 11 152700 175000.0 -22300.0 Residential Single Family 92 HUTCHINSON Andover 70024 2007 2008 7 188600 260000.0 -71400.0 Residential Single Family 93 LAKESIDE DR Andover 80013 2008 2008 10 10500 40250.0 -29750.0 Vacant Land NA LOT 12-1 BUNKER HILL RD Andover 800063 2008 2008 12 148300 149900.0 -1600.0 Residential Two Family 114-116 MYRTLE AVE Ansonia 80013 2008 2008 10 146000 17000.0 129000.0 Residential Two Family 114 TREMONT ST Ansonia 800061 2008 2008 12 143500 90000.0 53500.0 Residential Two Family 13-15 EAGLE ST Ansonia 80006 2008 2008 10 131700 188000.0 -56300.0 Residential Single Family 13 HIGHLAND TERR Ansonia 800051 2008 2008 12 173400 265000.0 -91600.0 Residential Single Family 136 BENZ ST Ansonia 800077 2008 2008 12 173400 236000.0 -62600.0 Residential Single Family 136 BENZ ST Ansonia 800086 2008 2008 11 164600 250000.0 -85400.0 Residential Single Family 136 HILL ST Ansonia 80016 2008 2008 10 205000 205000.0 0.0 Residential Three Family 139-141 N STATE ST Ansonia 80003 2008 2008

#### 4.5.  Analysis: Coordinate Plot in the year 2008

In [0]:
# Here we filter the dataset to only 2018, and select just the three columns showing address and sale amount
df2008 = df.filter(df.SaleYear  == 2008).select(['Address','Town','SaleAmount'])

# Join both the address and Town columns
df2008 = df2008.withColumn('FullAddress', F.concat(F.col('Address'), F.lit(' '),F.col('Town'))).drop('Address','Town')  \
                                                                           .select('FullAddress','SaleAmount')

print(df2008.count())
df2008.show(5)


32715
+--------------------+----------+
 FullAddress|SaleAmount|
+--------------------+----------+
6 FIELDSTONE LN B...| 508955.0|
111 LAKESIDE DR A...| 350000.0|
 12 ROSE LN Andover| 172500.0|
146 LAKESIDE DR A...| 239000.0|
15 DOGWOOD AVENUE...| 250000.0|
+--------------------+----------+
only showing top 5 rows

In [0]:
# Use databrick's display() command to download the first 1000 rows as a csv
df2008.display()

FullAddress SaleAmount 6 FIELDSTONE LN Beacon Falls 508955.0 111 LAKESIDE DR Andover 350000.0 12 ROSE LN Andover 172500.0 146 LAKESIDE DR Andover 239000.0 15 DOGWOOD AVENUE Andover 250000.0 177 SHODDY MILL RD Andover 325000.0 18 LAKESIDE DR Andover 224900.0 19 PARKER BRIDGE ROAD Andover 210000.0 2 MERRITT VALLEY ROAD Andover 221000.0 233 LAKE ROAD Andover 90000.0 242 GILEAD RD Andover 365000.0 25 ASPINALL DRIVE Andover 170000.0 29 STANLEY DR Andover 407000.0 30 ROUTE 6 Andover 179770.0 30 ROUTE 6 Andover 102000.0 321 LAKE RD Andover 180000.0 327 LAKE ROAD Andover 195500.0 33 GILEAD ROAD Andover 175000.0 336 HEBRON RD Andover 382500.0 343 LAKE RD Andover 285000.0 36 OAK FARMS RD Andover 595000.0 371 LAKE ROAD & LOT 5 LAKE ROA Andover 150000.0 4 SHADBLOW LN Andover 391500.0 44 WALES RD Andover 260000.0 5 SHADBLOW LN Andover 420000.0 51 WINDRUSH LN Andover 231000.0 587 ROUTE 6 Andover 170000.0 61 SHODDY MILL ROAD Andover 250000.0 76 WEST ST Andover 213750.0 85 GILEAD RD Andover 306200.0 901 EAST STREET Andover 270600.0 92 HUTCHINSON Andover 175000.0 93 LAKESIDE DR Andover 260000.0 LOT 12-1 BUNKER HILL RD Andover 40250.0 114-116 MYRTLE AVE Ansonia 149900.0 114 TREMONT ST Ansonia 17000.0 13-15 EAGLE ST Ansonia 90000.0 13 HIGHLAND TERR Ansonia 188000.0 136 BENZ ST Ansonia 265000.0 136 BENZ ST Ansonia 236000.0 136 HILL ST Ansonia 250000.0 139-141 N STATE ST Ansonia 205000.0 15 BERKSHIRE CR Ansonia 230000.0 15 CHESTER ST Ansonia 249900.0 16 FIFTH ST Ansonia 121000.0 16 GEORGE ST Ansonia 199000.0 16 HALL ST Ansonia 133000.0 16 HOTCHKISS TERR Ansonia 180000.0 16 ORCHARD ST Ansonia 125000.0 164 N STATE ST Ansonia 85000.0 17 BELLEVIEW TERR Ansonia 156500.0 17 LOCKE ST Ansonia 231959.0 20 HOLBROOK COURT Ansonia 170000.0 21 GRANITE TERR Ansonia 390000.0 213 PRINDLE AVE Ansonia 137000.0 222 NORTH STATE ST Ansonia 220000.0 24 SILVER HILL RD Ansonia 105000.0 3 MACINTOSH LA Ansonia 190000.0 32 MOULTHROP ST Ansonia 250000.0 35 YORK ST & 35 YORK ST REAR Ansonia 142500.0 37-39 CHURCH ST Ansonia 259700.0 38 CONDON DR Ansonia 380000.0 40 PRINDLE AVE Ansonia 295000.0 46 PRINDLE AVE Ansonia 215000.0 47 CHURCH ST Ansonia 167500.0 5 LINDGREN TERR Ansonia 250000.0 52 FORD ST Ansonia 345000.0 55 MORNINGSIDE DR Ansonia 190000.0 60 HODGE AVE Ansonia 230000.0 62 GREAT HILL RD Ansonia 315000.0 64 WOODLAWN AVE REAR Ansonia 213000.0 8 HALLIHAN LA Ansonia 297000.0 8 RESERVOIR DR Ansonia 237500.0 91 HODGE AVE Ansonia 287500.0 RO MORNINGSIDE DR Ansonia 182000.0 116 PUMPKIN HILL RD Ashford 210000.0 149 ASHFORD CENTER RD Ashford 3850000.0 157 OAKES RD Ashford 312500.0 181 SECKAR RD Ashford 265000.0 237 HILLSIDE RD Ashford 430000.0 274 FERENCE RD Ashford 129400.0 30 SUPINA RD Ashford 152000.0 32-8 POMPEY RD Ashford 76000.0 34 POMPEY HOLLOW RD Ashford 60000.0 394 FERENCE RD Ashford 281000.0 6 VIEW POINT DR Ashford 240000.0 72 LAUREL LANE Ashford 211000.0 91 MANSFIELD RD Ashford 145000.0 1 BLANCHARD RD Avon 259900.0 1 EAGLES GLEN Avon 595000.0 1 FARMSTEAD LN Avon 575000.0 1 MORGAN PL Avon 390000.0 1 MUIRFIELD LN Avon 560000.0 1 SPRUCEWOOD LN Avon 280000.0 10 BURNHAM HOLLOW Avon 505000.0 10 HERITAGE DR Avon 160000.0 10 WYNDEMERE Avon 587000.0 100 HIGH RIDGE RD Avon 399900.0 102 FOX DEN RD Avon 385000.0 103 ROSEWOOD RD Avon 390000.0 106 FOX HOLLOW Avon 384000.0 109 REVERKNOLLS Avon 608000.0 11 ALEXANDRA LN Avon 455000.0 11 BILTMORE CT Avon 735000.0 11 BYRON DR Avon 340000.0 11 GARNET HILL LN Avon 1200000.0 11 QUAIL RIDGE DR Avon 695000.0 11 SWEET BRIAR LN Avon 278000.0 110 WEST AVON RD Avon 580000.0 115 FOX HOLLOW Avon 415000.0 118 JUNIPER DR Avon 375000.0 12 BLUESTONE LN Avon 2500.0 12 BLUESTONE LN Avon 791500.0 12 CLOVER CT Avon 232500.0 12 CONIFER LN Avon 199900.0 12 LAKEVIEW BLVD Avon 210000.0 12 LAUREL LN Avon 185000.0 12 STONEFIELD RD Avon 1437500.0 12 STRATHMORE LN Avon 275000.0 122 ANVIL DR Avon 357000.0 124 TAMARA CIR Avon 160000.0 13 FINCH RUN Avon 210000.0 133 LOFGREN RD Avon 2125000.0 135 JUNIPER DR Avon 390000.0 139 MALLARD DR Avon 218000.0 14 ANVIL DR Avon 

#### Import Geocoded address CSV into the notebook

In [0]:
path = "dbfs:/FileStore/shared_uploads/kariboalap@uni.coventry.ac.uk/Connecticut_Sample_2008___Sheet1.csv"

geodf =spark.read.format("csv").load(path, inferSchema=True, header=True)

In [0]:
geodf.display()

GeocodedAddress Latitude Longitude SaleAmount Status Fieldstone Lane, Beacon Falls, CT, USA 41.425483 -73.040236 508955 doubt 111 Lakeside Dr, Andover, CT, USA 41.720633 -72.361915 350000 success 12 Rose Ln, Andover, CT, USA 41.71064 -72.357205 172500 success 146 Lakeside Dr, Andover, CT, USA 41.72311 -72.362294 239000 success Andover, England, United Kingdom 51.214307 -1.478605 250000 doubt 177 Shoddy Mill Rd, Andover, CT, USA 41.739405 -72.403346 325000 success 18 Lakeside Dr, Andover, CT, USA 41.713959 -72.360867 224900 success 19 Parker Bridge Rd, Andover, CT, USA 41.732249 -72.336445 210000 success 2 Merritt Valley Rd, Andover, CT, USA 41.73513 -72.36166 221000 success 233 Lake Rd, Andover, CT, USA 41.710886 -72.358037 90000 success 242 Gilead Rd, Andover, CT, USA 41.701641 -72.39123 365000 success 25 Aspinall Dr, Andover, CT, USA 41.751208 -72.397896 170000 success 29 Stanley Dr, Andover, CT, USA 41.760837 -72.38316 407000 success 30 Route 6, Andover, CT, USA 41.762327 -72.405356 179770 success 30 Route 6, Andover, CT, USA 41.762327 -72.405356 102000 success 321 Lake Rd, Andover, CT, USA 41.714861 -72.352313 180000 success 327 Lake Rd, Andover, CT, USA 41.714681 -72.351663 195500 success 33 Gilead Rd, Andover, CT, USA 41.718709 -72.375832 175000 success 336 Hebron Rd, Andover, CT, USA 41.707114 -72.367312 382500 success 343 Lake Rd, Andover, CT, USA 41.715636 -72.350014 285000 success 36 Oak Farms Rd, Andover, CT, USA 41.728868 -72.408541 595000 success Andover, England, United Kingdom 51.214307 -1.478605 150000 doubt Shadblow Lane, Andover, CT, USA 41.754405 -72.372 391500 doubt 44 Wales Rd, Andover, CT, USA 41.734431 -72.389351 260000 success Shadblow Lane, Andover, CT, USA 41.754405 -72.372 420000 doubt 51 Windrush Ln, Andover, CT, USA 41.701714 -72.383453 231000 success 587 Route 6, Andover, CT, USA 41.728019 -72.347744 170000 success 61 Shoddy Mill Rd, Andover, CT, USA 41.739619 -72.388994 250000 success 76 West St, Andover, CT, USA 41.707963 -72.35538 213750 success 85 Gilead Rd, Andover, CT, USA 41.715806 -72.380052 306200 success 901 East St, Andover, CT, USA 41.728975 -72.419956 270600 success Andover, England, United Kingdom 51.214307 -1.478605 175000 doubt 93 Lakeside Dr, Andover, CT, USA 41.719355 -72.3613 260000 success Andover, England, United Kingdom 51.214307 -1.478605 40250 doubt Myrtle Avenue, Ansonia, CT, USA 41.344691 -73.068816 149900 doubt 114 Tremont St, Ansonia, CT, USA 41.34213 -73.074731 17000 success Eagle Street, Ansonia, CT, USA 41.34597 -73.070507 90000 doubt Highland Terrace, Ansonia, CT, USA 41.351224 -73.077565 188000 doubt 136 Benz St, Ansonia, CT, USA 41.331917 -73.053134 265000 success 136 Benz St, Ansonia, CT, USA 41.331917 -73.053134 236000 success 136 Hill St, Ansonia, CT, USA 41.341407 -73.065828 250000 success North State Street, Ansonia, CT, USA 41.352383 -73.077559 205000 doubt Ansonia, CT, USA 41.343318 -73.066358 230000 doubt 15 Chester St, Ansonia, CT, USA 41.333584 -73.044761 249900 success 16 Fifth St, Ansonia, CT, USA 41.35303 -73.08267 121000 success 16 George St, Ansonia, CT, USA 41.348006 -73.073128 199000 success 16 Hall St, Ansonia, CT, USA 41.340711 -73.08894 133000 success Hotchkiss Terrace, Ansonia, CT, USA 41.361277 -73.086346 180000 doubt 16 Orchard St, Ansonia, CT, USA 41.342401 -73.071227 125000 success 164 N State St, Ansonia, CT, USA 41.351779 -73.076041 85000 success Belleview Terrace, Ansonia, CT, USA 41.337777 -73.066184 156500 doubt 17 Locke St, Ansonia, CT, USA 41.351609 -73.075447 231959 success 20 Holbrook Ct, Ansonia, CT, USA 41.345489 -73.087694 170000 success Granite Terrace, Ansonia, CT, USA 41.355731 -73.076893 390000 doubt 213 Prindle Ave, Ansonia, CT, USA 41.338269 -73.06396 137000 success 222 N State St, Ansonia, CT, USA 41.352987 -73.078579 220000 success 24 Silver Hill Rd, Ansonia, CT, USA 41.344364 -73.097706 105000 success Ansonia, CT, USA 41.343318 -73.066358 190000 doubt 32 Moulthrop St, Ansonia, CT, USA 41.346892 -73.073777 250000 succes

In [0]:
from pyspark.sql.types import DoubleType

# Select only successful geocoding attempts
geodf_success = geodf.filter(geodf.Status  == "success").select(['Latitude','Longitude','SaleAmount'])

# Cast the coordinate columns to numerical datatypes
geodf_success = geodf_success.withColumn("Longitude", geodf_success["Longitude"].cast(DoubleType()))
geodf_success = geodf_success.withColumn("Latitude", geodf_success["Latitude"].cast(DoubleType()))

geodf_success.show(5)

+---------+----------+----------+
 Latitude| Longitude|SaleAmount|
+---------+----------+----------+
41.720633|-72.361915| 350000|
 41.71064|-72.357205| 172500|
 41.72311|-72.362294| 239000|
41.739405|-72.403346| 325000|
41.713959|-72.360867| 224900|
+---------+----------+----------+
only showing top 5 rows

In [0]:
import pandas as pd
import matplotlib.pyplot as plt 
from bokeh.resources import CDN
from bokeh.embed import components
from bokeh.embed import file_html
from bokeh.io import show
from bokeh.plotting import figure
from bokeh.models import WMTSTileSource
from bokeh.models import LinearColorMapper, ColorBar, ColumnDataSource
from bokeh.transform import linear_cmap
from bokeh.palettes import Cividis, Spectral6

In [0]:
import numpy as np

# Function to convert coordinate data to mercator format
def to_web_mercator(data, longitude="Longitude", latitude="Latitude"):
    
    k = 6378137
    data["new_long"] = data[longitude] * (k * np.pi/180.0)
    data["new_lat"] = np.log(np.tan((90 + data[latitude]) * np.pi/360.0)) * k
    return data


map_df = geodf_success.toPandas()
converted_df = to_web_mercator(map_df)

In [0]:
converted_df.head()

Latitude 
 Longitude 
 SaleAmount 
 new_long 
 new_lat 
 
 
 
 
 0 
 41.720633 
 -72.361915 
 350000 
 -8.055292e+06 
 5.119223e+06 
 
 
 1 
 41.710640 
 -72.357205 
 172500 
 -8.054767e+06 
 5.117733e+06 
 
 
 2 
 41.723110 
 -72.362294 
 239000 
 -8.055334e+06 
 5.119592e+06 
 
 
 3 
 41.739405 
 -72.403346 
 325000 
 -8.059904e+06 
 5.122023e+06 
 
 
 4 
 41.713959 
 -72.360867 
 224900 
 -8.055175e+06 
 5.118228e+06

In [0]:
url = 'http://a.basemaps.cartocdn.com/rastertiles/voyager/{Z}/{X}/{Y}.png'
attribution = "Tiles by Carto, under CC BY 3.0. Data by OSM, under ODbL"

# web mercator coordinates
USA = x_r,y_r = ((-13884029,-7453304), (2698291,6455972))

final_map = figure(tools='pan, wheel_zoom', x_range=x_r, y_range=y_r, x_axis_type="mercator", y_axis_type="mercator", )
           
# Attribution
final_map.add_tile(WMTSTileSource(url=url, attribution=attribution))

# Colour mapper
new_mapper = linear_cmap(field_name='z', palette=Spectral6 ,low=min(converted_df.SaleAmount)*3, high=max(converted_df.SaleAmount)/2)

source = ColumnDataSource(dict(x=converted_df['new_long'],y=converted_df['new_lat'], z=converted_df['SaleAmount']))

# Bokeh plot
final_map.circle(x='x', y='y', line_color=new_mapper,color=new_mapper, fill_alpha=1, size=10, source=source)

# Colour Bar
c_bar = ColorBar(color_mapper=new_mapper['transform'], width=15,  location=(0,0))
final_map.add_layout(c_bar, 'right')

html = file_html(final_map, CDN, "Connecticut")

# display the figure
displayHTML(html)   

<!DOCTYPE html>
 
 
 
 
 
 Connecticut

## 5  Machine Learning

#### 5.1. Dataframe preparation

In [0]:
mldf = df
mldf.display()

SerialNumber ListYear SaleYear SaleMonth AssessedValue SaleAmount PriceDifference PropertyType ResidentialType Address Town 40043 2004 2005 5 850 50000.0 -49150.0 Vacant Land NA BURNAP BRK RD Andover 30070 2003 2004 8 157300 175000.0 -17700.0 Vacant Land NA ROUTE 6 Andover 40036 2004 2005 4 22400 20000.0 2400.0 Vacant Land NA ROUTE 6 Andover 50249 2005 2006 7 0 197400.0 -197400.0 Condo NA 13 MCINTOSH LN Ansonia 50125 2005 2006 2 331100 375000.0 -43900.0 Condo NA 22 WESTFIELD AVE 5 & Ansonia 50059 2005 2006 5 1196070 3700000.0 -2503930.0 Apartments NA 52 PERRY HL RD Ashford 140042 2014 2015 7 3900 2500.0 1400.0 Vacant Land NA ASHFORD LAKE DR Ashford 900009 2009 2009 12 64000 160000.0 -96000.0 Vacant Land NA HILLSIDE RD Ashford 20384 2002 2003 6 334750 577500.0 -242750.0 Condo NA 1 WILLS WALK Avon 120148 2012 2013 3 125870 125000.0 870.0 Commercial NA 11 CHESTNUT DRIVE Avon 40478 2004 2005 7 109550 229500.0 -119950.0 Condo NA 11 CLOVER CT Avon 20287 2002 2003 4 77000 147500.0 -70500.0 Condo NA 11 WILLOW LN Avon 20345 2002 2003 5 38500 546987.0 -508487.0 Condo NA 12 WYNDEMERE Avon 100064 2010 2010 12 354310 550000.0 -195690.0 Condo NA 16 GATEWOOD Avon 40056 2004 2004 11 105350 175000.0 -69650.0 Condo NA 2 DOVE CIR Avon 60014 2006 2006 10 102900 250000.0 -147100.0 Condo NA 2 SYCAMORE LN Avon 40163 2004 2005 1 307020 430000.0 -122980.0 Condo NA 2 TEMPLETON CT Avon 20194 2002 2003 2 173880 380000.0 -206120.0 Condo NA 3 BRIDLE PATH Avon 10408 2001 2002 7 87080 165000.0 -77920.0 Condo NA 4 SYCAMORE LN Avon 20207 2002 2003 2 125860 235000.0 -109140.0 Condo NA 48 BYRON DR Avon 10340 2001 2002 6 42000 10000.0 32000.0 Vacant Land NA 50 SCHOOL ST Avon 50288 2005 2006 6 170790 339500.0 -168710.0 Condo NA 6 GREENWICH LN Avon 50257 2005 2006 5 153240 284000.0 -130760.0 Condo NA 7 ESSEX CT Avon 20236 2002 2003 3 163910 327500.0 -163590.0 Condo NA 7 MADISON LN Avon 40404 2004 2005 6 189700 350000.0 -160300.0 Condo NA 7 MADISON LN Avon 40384 2004 2005 6 233180 399000.0 -165820.0 Condo NA 70 FOX HOLW Avon 20448 2002 2003 7 87080 189000.0 -101920.0 Condo NA 9 WILLOW LN Avon 30284 2003 2004 6 13600 245725.0 -232125.0 Condo NA 92 HAWTHORNE DR Berlin 70072 2007 2008 5 0 508955.0 -508955.0 Condo NA 6 FIELDSTONE LN Beacon Falls 50320 2005 2006 8 183700 362000.0 -178300.0 Condo NA 159 SCARBOROUGH FARE Berlin 900194 2009 2010 6 0 362031.0 -362031.0 Condo NA 17 IRIS WAY Berlin 100185 2010 2011 6 413300 50000.0 363300.0 Vacant Land NA 501 BERLIN TPKE Berlin 19010 2001 2001 11 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 20076 2002 2003 3 149840 0.0 149840.0 Residential Single Family 29 TAWNEY DR Bethany 50055 2005 2006 2 106580 245000.0 -138420.0 Vacant Land NA 51 JUDD HL RD LT 7 Bethany 10134 2001 2002 6 171350 0.0 171350.0 Residential Single Family 66 TUTTLE CT Bethany 40109 2004 2004 12 121590 244000.0 -122410.0 Condo NA 10 IDLEWOOD Bethel 50053 2005 2005 11 102060 219000.0 -116940.0 Condo NA 20 JUNIPER RD Bethel 60265 2006 2007 8 102060 180000.0 -77940.0 Condo NA 20 JUNIPER RD Bethel 50252 2005 2006 6 123620 280000.0 -156380.0 Condo NA 23 DEER RUN UT 56 Bethel 50212 2005 2006 5 183630 550000.0 -366370.0 Vacant Land NA 28-38 TURKEY PLN RD Bethel 100012 2010 2010 10 142310 160000.0 -17690.0 Condo NA 2B BEACH STREET Bethel 50216 2005 2006 6 128450 260000.0 -131550.0 Condo NA 38 IDLEWOOD UT 122 Bethel 50285 2005 2006 8 184870 328400.0 -143530.0 Condo NA 38 KINGSWOOD DR UT 60 Bethel 30325 2003 2004 6 0 350000.0 -350000.0 Vacant Land NA 7 SARA'S WAY Bethel 50068 2005 2005 12 88200 181000.0 -92800.0 Condo NA 75 JUNIPER RD Bethel 90078 2009 2010 1 124530 140000.0 -15470.0 Condo NA 75 JUNIPER ROAD Bethel 40412 2004 2005 9 67200 96000.0 -28800.0 Commercial NA 7D SCHOOL ST Bethel 100092 2010 2011 2 141050 155000.0 -13950.0 Condo NA 8-O BEACH STREET Bethel 50015 2005 2005 12 83800 180000.0 -96200.0 Vacant Land NA HICKORY LN Bethlehem 50278 2005 2006 4 515250 850000.0 -334750.0 Residential NA 11 MOUNTAIN AVE Bloomfield 50010 2005 2005 10 22765

For the Machine Learning task, we can drop the `address` column since it does not information critical to our model, and would potentially create too many categories during one-hot encoding.

In [0]:
mldf = mldf.drop('Address')

##### Firstly, the categorical columns need to be converted to numerical types before being fed into any Machine Learning model

In [0]:
from pyspark.ml import Pipeline
from pyspark.ml.feature import VectorAssembler, OneHotEncoder, StringIndexer

categorical_columns = ["PropertyType", "ResidentialType", "Town"]
stage = [] # The steps of the pipeline 


for cat_col in categorical_columns:
    strIndexer = StringIndexer(inputCol=cat_col, outputCol=cat_col + "Index")
    encoder = OneHotEncoder(inputCols=[strIndexer.getOutputCol()], outputCols=[cat_col + "classVector"])
                                     
    stage += [strIndexer, encoder]

numericals = ["ListYear","SaleYear","SaleMonth","AssessedValue","SaleAmount"]    
assembler_in = [col + "classVector" for col in categorical_columns] + numericals

ass1 = VectorAssembler(inputCols=assembler_in, outputCol="features")
stage += [ass1]

##### Next we make the pipeline

In [0]:
# Build Pipeline.
pl = Pipeline(stages=stage)
pl_model = pl.fit(mldf)
transformed_df = pl_model.transform(mldf)

In [0]:
# Pick only the predictor and target columns
processed_df = transformed_df.select(['SaleAmount','features'])

In [0]:
processed_df.show(5)

+----------+--------------------+
SaleAmount| features|
+----------+--------------------+
 50000.0|(186,[2,9,162,181...|
 175000.0|(186,[2,9,162,181...|
 20000.0|(186,[2,9,162,181...|
 197400.0|(186,[1,9,98,181,...|
 375000.0|(186,[1,9,98,181,...|
+----------+--------------------+
only showing top 5 rows

#### 5.2. Model Construction

In [0]:
# Split data into training and testing portions

train, test = processed_df.randomSplit([0.7,0.3], seed=7)
train.describe().show()
test.describe().show()

+-------+------------------+
summary| SaleAmount|
+-------+------------------+
 count| 548761|
 mean|360230.30608386896|
 stddev|2202676.6968459864|
 min| 0.0|
 max| 9.4094E8|
+-------+------------------+

+-------+------------------+
summary| SaleAmount|
+-------+------------------+
 count| 236059|
 mean| 355686.3612122393|
 stddev|1594513.3718408796|
 min| 0.0|
 max| 3.955E8|
+-------+------------------+

In [0]:
from pyspark.ml.regression import LinearRegression, DecisionTreeRegressor
from pyspark.ml.evaluation import RegressionEvaluator

#### Linear Regression

In [0]:
sale_lr = LinearRegression(featuresCol="features", 
                           labelCol="SaleAmount")

sale_trained_model_lr = sale_lr.fit(train)
predictions = sale_trained_model_lr.transform(test)

eval1 = RegressionEvaluator(labelCol="SaleAmount", 
                            predictionCol="prediction", 
                            metricName="rmse")
rmse = eval1.evaluate(predictions)

eval2 = RegressionEvaluator(labelCol="SaleAmount", 
                            predictionCol="prediction", 
                            metricName="mae")
mae = eval2.evaluate(predictions)

eval3 = RegressionEvaluator(labelCol="SaleAmount", 
                            predictionCol="prediction", 
                            metricName="r2")
r2 = eval3.evaluate(predictions)

print(f"Root Mean Squared Error = {rmse}")
print(f"Mean Absolute Error = {mae}")
print(f"R-Squared Value (R^2) = {r2}")

Root Mean Squared Error = 2.6049097527823197e-10
Mean Absolute Error = 3.145402900017911e-11
R-Squared Value (R^2) = 1.0

#### Reality Check on what is an absurdly good result.

In [0]:
from pyspark.sql.functions import corr

mldf.select(corr("AssessedValue","SaleAmount")).show()

+-------------------------------+
corr(AssessedValue, SaleAmount)|
+-------------------------------+
 0.2589125133226587|
+-------------------------------+

#### Decision Tree Regression

In [0]:
sale_dtr = DecisionTreeRegressor(featuresCol="features", 
                                 labelCol="SaleAmount")
sale_trained_model_dtr = sale_dtr.fit(train)
predictions = sale_trained_model_dtr.transform(test)

eval1 = RegressionEvaluator(labelCol="SaleAmount", 
                            predictionCol="prediction", 
                            metricName="rmse")
rmse = eval1.evaluate(predictions)

eval2 = RegressionEvaluator(labelCol="SaleAmount", 
                            predictionCol="prediction", 
                            metricName="mae")
mae = eval2.evaluate(predictions)

evalr3 = RegressionEvaluator(labelCol="SaleAmount", 
                             predictionCol="prediction", 
                             metricName="r2")
r2 = eval3.evaluate(predictions)

print(f"Root Mean Squared Error = {rmse}")
print(f"Mean Absolute Error = {mae}")
print(f"R-Squared Value (R^2) = {r2}")

Root Mean Squared Error = 1197821.3730674104
Mean Absolute Error = 60649.32847053634
R-Squared Value (R^2) = 0.43567456557560436